---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 03-basics-factors-dataframes.md in _episodes_rmd/
title: "R Basics continued - factors and data frames"
teaching: 45
exercises: 20
questions:
- "How do I get started with tabular data (e.g. spreadsheets) in R?"
- "What are some best practices for reading data into R?"
- "How do I save tabular data generated in R?"
objectives:
- "Load external data from a .csv file into a data frame."
- "Describe what a data frame is."
- "Summarize the contents of a data frame."
- "Use indexing to subset specific portions of data frames."
- "Describe what a factor is."
- "Convert between strings and factors."
- "Reorder and rename factors."
- "Change how character strings are handled in a data frame."
- "Format dates."
keypoints:
- "It is easy to import data into R from tabular formats including Excel.
  However, you still need to check that R has imported and interpreted your
  data correctly"
- "There are best practices for organizing your data (keeping it tidy) and R
  is great for this"
- "Base R has many useful functions for manipulating your data, but all of R's
  capabilities are greatly enhanced by software packages developed by the
  community"
source: Rmd
---



## Loading the samples data

We are investigating gene expression in mice and want to keep track of animals
included in different experiments. The dataset is stored as a comma separated
value (CSV) file. Each row holds information for a single animal, and the
columns represent:

| Column                | Description                                                                                         |
|-----------------------|-----------------------------------------------------------------------------------------------------|
| animal\_id            | 6-digit code for the animal                                                                         |
| month                 | month of experiment                                                                                 |
| day                   | day of experiment                                                                                   |
| year                  | year of experiment                                                                                  |                                                                |
| mouse\_line           | short code for the mouse line                                                                       |
| strain                | background strain of the mouse line                                                                 |
| developmental\_stage  | stage of development (e.g. "embryo", "pup" or "adult")                                              |
| sex                   | "F" (female), "M" (male) or NA (unknown)                                                            |
| organism\_part        | part of the animal used for the experiment                                                          |
| genotype              | "wild type" (2 wild type alleles), "mutant" (1 mutated allele) or "knockout" (both alleles mutated) |
| experiment\_type      | kind of experiment; "IHC" for immunnohistochemistry                                                 |
| researcher            | researcher responsible for the experiment                                                           |

### Downloading the data

We are going to use the R function `download.file()` to download the CSV file
that contains data from Figshare, and we will use `read_csv()` to load the
content of the CSV file into R.

Inside the `download.file` command, the first entry is a character string with the
source URL ("https://nbisweden.github.io/module-r-intro-dm-practices/data/samples_r_lesson.csv"). 
This source URL downloads a CSV file from FitHub. The text after the comma
("data_raw/samples.csv") is the destination of the file on your local machine.
You'll need to have a folder on your machine called "data_raw" where 
you'll download the file. So this command downloads a file from the web, names 
it "samples.csv" and adds it to a preexisting folder named "data_raw".


~~~
download.file(
  url = "https://nbisweden.github.io/module-r-intro-dm-practices/data/samples_r_lesson.csv",
  destfile = "data_raw/samples_r_lesson.csv")
~~~
{: .language-r}


### Reading the data into R

The file has now been downloaded to the destination you specified, but R has not 
yet loaded the data from the file into memory. To do this, we can use the 
`read_csv()` function from the **`tidyverse`** package. 

Packages in R are basically sets of additional functions that let you do more
stuff. The functions we've been using so far, like `round()`, `sqrt()`, or
`c()`, come built into R; packages give you access to additional functions. 
Before you use a package for the first time you need to install it on your 
machine, and then you should import it in every subsequent R session when you 
need it. 

To install the **`tidyverse`** package, we can type
`install.packages("tidyverse")` straight into the console. In fact, it's better 
to write this in the console than in our script for any package, as there's no 
need to re-install packages every time we run the script. 
Then, to load the package type:


~~~
## load the tidyverse packages, incl. dplyr
library(tidyverse)
~~~
{: .language-r}

Now we can use the functions from the **`tidyverse`** package. Let's use
`read_csv()` to read the data into a data frame (we will learn more about data
frames later):


~~~
samples <- read_csv("data_raw/samples_r_lesson.csv")
~~~
{: .language-r}

~~~

── Column specification ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────
cols(
  animal_id = col_double(),
  month = col_double(),
  day = col_double(),
  year = col_double(),
  mouse_line = col_character(),
  strain = col_character(),
  developmental_stage = col_character(),
  sex = col_character(),
  organism_part = col_character(),
  genotype = col_character(),
  experiment_type = col_character(),
  researcher = col_character()
)
~~~
{: .output}

You will see the text `Column specification`, followed by each column name and
its data type. When you execute `read_csv` on a data file, it looks through the
first 1000 rows of each column and guesses its data type. For example, in this
dataset, `read_csv()` reads `animal_id` as `col_double` (a numeric data type),
and `mouse_line` as `col_character`. You have the option to specify the data
type for a column manually by using the `col_types` argument in `read_csv`.

We can see the contents of the first few lines of the data by typing its name:
`samples`. By default, this will show show you as many rows and columns of the
data as fit on your screen. If you wanted to the first 50 rows, you could type
`print(samples, n = 50)`. We can also extract the first few lines of this
data using the function `head()`:


~~~
head(samples)
~~~
{: .language-r}



~~~
# A tibble: 6 x 12
  animal_id month   day  year mouse_line strain  developmental_stage sex  
      <dbl> <dbl> <dbl> <dbl> <chr>      <chr>   <chr>               <chr>
1    800793     1     6  2018 Alk3       BALB/cJ adult               F    
2    804396     1     7  2019 Vegfr      C57BL/6 adult               M    
3    805431     1    12  2018 Vegfr      C57BL/6 pup                 F    
4    805992     1    13  2019 Vegfr      C57BL/6 pup                 M    
5    808935     1    14  2020 Alk3       BALB/cJ adult               M    
6    810875     1    16  2019 Alk6       C57BL/6 embryo              <NA> 
# … with 4 more variables: organism_part <chr>, genotype <chr>,
#   experiment_type <chr>, researcher <chr>
~~~
{: .output}

Unlike the `print()` function, `head()` returns the extracted data. You could 
use it to assign the first 20 rows of `samples` to an object using
`samples_head20 <- head(samples, 20)`. This can be useful if you want to try 
out complex computations on a subset of your data before you apply them to the
whole data set. There is a similar function that lets you extract the last few
lines of the dataset. It is called (you might have guessed it) `tail()`.

To open the dataset in RStudio's Data Viewer, use the `view()` function:


~~~
view(samples)
~~~
{: .language-r}

> ### Note
>
> `read_csv()` assumes that fields are delineated by commas, however, in several
> countries, the comma is used as a decimal separator and the semicolon (;) is
> used as a field delineator. If you want to read in this type of files in R,
> you can use the `read_csv2()` function. It behaves like `read_csv()` but
> uses different parameters for the decimal and the field separators. 
> There is also the `read_tsv()` for tab separated data files and `read_delim()` 
> for less common formats.
> Check out the help for `read_csv()` by typing `?read_csv` to learn more. 
>
> In addition to the above versions of the csv format, you should develop the
> habits of looking at and record some parameters of your csv files. For
> instance, the character encoding, control characters used for line ending,
> date format (if the date is not splitted into three variables), and the
> presence of unexpected [newlines](https://en.wikipedia.org/wiki/Newline) are
> important characteristics of your data files. Those parameters will ease up
> the import step of your data in R.

## What are data frames?

When we loaded the data into R, it got stored as an object of class `tibble`, 
which is a special kind of data frame (the difference is not important for our 
purposes, but you can learn more about tibbles [here](https://tibble.tidyverse.org/)). 
Data frames are the _de facto_ data structure for most tabular data, and what we
use for statistics and plotting. Data frames can be created by hand, but most
commonly they are generated by functions like `read_csv()`; in other words, when
importing spreadsheets from your hard drive or the web.

A data frame is the representation of data in the format of a table where the
columns are vectors that all have the same length. Because columns are vectors,
each column must contain a single type of data (e.g., characters, integers,
factors). For example, here is a figure depicting a data frame comprising a
numeric, a character, and a logical vector.

![](../fig/data-frame.svg)
We can see this also when inspecting the <b>str</b>ucture of a data frame
with the function `str()`:


~~~
str(samples)
~~~
{: .language-r}



~~~
spec_tbl_df[,12] [100 × 12] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ animal_id          : num [1:100] 800793 804396 805431 805992 808935 ...
 $ month              : num [1:100] 1 1 1 1 1 1 1 1 1 1 ...
 $ day                : num [1:100] 6 7 12 13 14 16 19 19 20 24 ...
 $ year               : num [1:100] 2018 2019 2018 2019 2020 ...
 $ mouse_line         : chr [1:100] "Alk3" "Vegfr" "Vegfr" "Vegfr" ...
 $ strain             : chr [1:100] "BALB/cJ" "C57BL/6" "C57BL/6" "C57BL/6" ...
 $ developmental_stage: chr [1:100] "adult" "adult" "pup" "pup" ...
 $ sex                : chr [1:100] "F" "M" "F" "M" ...
 $ organism_part      : chr [1:100] NA "lung" "lung" "lung" ...
 $ genotype           : chr [1:100] "wild type" "wild type" "wild type" "wild type" ...
 $ experiment_type    : chr [1:100] "behaviour" "sequencing assay" "IHC" "IHC" ...
 $ researcher         : chr [1:100] "Kim" "Sam" "Sam" "Sam" ...
 - attr(*, "spec")=
  .. cols(
  ..   animal_id = col_double(),
  ..   month = col_double(),
  ..   day = col_double(),
  ..   year = col_double(),
  ..   mouse_line = col_character(),
  ..   strain = col_character(),
  ..   developmental_stage = col_character(),
  ..   sex = col_character(),
  ..   organism_part = col_character(),
  ..   genotype = col_character(),
  ..   experiment_type = col_character(),
  ..   researcher = col_character()
  .. )
~~~
{: .output}

## Inspecting data frames

We already saw how the functions `head()` and `str()` can be useful to check the
content and the structure of a data frame. Here is a non-exhaustive list of
functions to get a sense of the content/structure of the data. Let's try them
out!

* Size:
    * `dim(samples)` - returns a vector with the number of rows in the first
      element, and the number of columns as the second element
      (the **dim**ensions of the object)
    * `nrow(samples)` - returns the number of rows
    * `ncol(samples)` - returns the number of columns

* Content:
    * `head(samples)` - shows the first 6 rows
    * `tail(samples)` - shows the last 6 rows

* Names:
    * `names(samples)` - returns the column names (synonym of `colnames()` for
      `data.frame` objects)
    * `rownames(samples)` - returns the row names

* Summary:
    * `str(samples)` - structure of the object and information about the class,
      length and content of each column
    * `summary(samples)` - summary statistics for each column

Note: most of these functions are "generic", they can be used on other types of
objects besides `data.frame`.


> ## Challenge
>
> Based on the output of `str(samples)`, can you answer the following questions?
>
> * What is the class of the object `samples`?
> * How many rows and how many columns are in this object?
>
>> ## Solution
>>
>> 
>> ~~~
>> str(samples)
>> ~~~
>> {: .language-r}
>> 
>> 
>> 
>> ~~~
>> spec_tbl_df[,12] [100 × 12] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
>>  $ animal_id          : num [1:100] 800793 804396 805431 805992 808935 ...
>>  $ month              : num [1:100] 1 1 1 1 1 1 1 1 1 1 ...
>>  $ day                : num [1:100] 6 7 12 13 14 16 19 19 20 24 ...
>>  $ year               : num [1:100] 2018 2019 2018 2019 2020 ...
>>  $ mouse_line         : chr [1:100] "Alk3" "Vegfr" "Vegfr" "Vegfr" ...
>>  $ strain             : chr [1:100] "BALB/cJ" "C57BL/6" "C57BL/6" "C57BL/6" ...
>>  $ developmental_stage: chr [1:100] "adult" "adult" "pup" "pup" ...
>>  $ sex                : chr [1:100] "F" "M" "F" "M" ...
>>  $ organism_part      : chr [1:100] NA "lung" "lung" "lung" ...
>>  $ genotype           : chr [1:100] "wild type" "wild type" "wild type" "wild type" ...
>>  $ experiment_type    : chr [1:100] "behaviour" "sequencing assay" "IHC" "IHC" ...
>>  $ researcher         : chr [1:100] "Kim" "Sam" "Sam" "Sam" ...
>>  - attr(*, "spec")=
>>   .. cols(
>>   ..   animal_id = col_double(),
>>   ..   month = col_double(),
>>   ..   day = col_double(),
>>   ..   year = col_double(),
>>   ..   mouse_line = col_character(),
>>   ..   strain = col_character(),
>>   ..   developmental_stage = col_character(),
>>   ..   sex = col_character(),
>>   ..   organism_part = col_character(),
>>   ..   genotype = col_character(),
>>   ..   experiment_type = col_character(),
>>   ..   researcher = col_character()
>>   .. )
>> ~~~
>> {: .output}
>> * The object `samples` is of class `data.frame`, or more specifically a
>>   `tibble` (`spec_tbl_df/tbl_df/tbl/data.frame`)
>> * Rows and columns: 100 rows and 12 columns
> {: .solution}
{: .challenge}


## Indexing and subsetting data frames

Our samples data frame has rows and columns (it has 2 dimensions), if we want to
extract some specific data from it, we need to specify the "coordinates" we
want from it. Row numbers come first, followed by column numbers.


~~~
# first element in the first column of the data frame
samples[1, 1]   
~~~
{: .language-r}



~~~
# A tibble: 1 x 1
  animal_id
      <dbl>
1    800793
~~~
{: .output}



~~~
# first element in the 6th column
samples[1, 6]   
~~~
{: .language-r}



~~~
# A tibble: 1 x 1
  strain 
  <chr>  
1 BALB/cJ
~~~
{: .output}



~~~
# first column of the data frame
samples[, 1]    
~~~
{: .language-r}



~~~
# A tibble: 100 x 1
   animal_id
       <dbl>
 1    800793
 2    804396
 3    805431
 4    805992
 5    808935
 6    810875
 7    812308
 8    814334
 9    816649
10    819947
# … with 90 more rows
~~~
{: .output}



~~~
# first column of the data frame
samples[1]      
~~~
{: .language-r}



~~~
# A tibble: 100 x 1
   animal_id
       <dbl>
 1    800793
 2    804396
 3    805431
 4    805992
 5    808935
 6    810875
 7    812308
 8    814334
 9    816649
10    819947
# … with 90 more rows
~~~
{: .output}



~~~
# first three rows of the 6th column
samples[1:3, 6] 
~~~
{: .language-r}



~~~
# A tibble: 3 x 1
  strain 
  <chr>  
1 BALB/cJ
2 C57BL/6
3 C57BL/6
~~~
{: .output}



~~~
# the 3rd row of the data frame
samples[3, ]    
~~~
{: .language-r}



~~~
# A tibble: 1 x 12
  animal_id month   day  year mouse_line strain  developmental_stage sex  
      <dbl> <dbl> <dbl> <dbl> <chr>      <chr>   <chr>               <chr>
1    805431     1    12  2018 Vegfr      C57BL/6 pup                 F    
# … with 4 more variables: organism_part <chr>, genotype <chr>,
#   experiment_type <chr>, researcher <chr>
~~~
{: .output}



~~~
# equivalent to head_samples <- head(samples)
head_samples <- samples[1:6, ] 
~~~
{: .language-r}

`:` is a special function that creates numeric vectors of integers in increasing
or decreasing order, test `1:10` and `10:1` for instance.

You can also exclude certain indices of a data frame using the "`-`" sign:


~~~
samples[, -1]       # the whole data frame, except the first column
~~~
{: .language-r}



~~~
# A tibble: 100 x 11
   month   day  year mouse_line strain  developmental_stage sex   organism_part
   <dbl> <dbl> <dbl> <chr>      <chr>   <chr>               <chr> <chr>        
 1     1     6  2018 Alk3       BALB/cJ adult               F     <NA>         
 2     1     7  2019 Vegfr      C57BL/6 adult               M     lung         
 3     1    12  2018 Vegfr      C57BL/6 pup                 F     lung         
 4     1    13  2019 Vegfr      C57BL/6 pup                 M     lung         
 5     1    14  2020 Alk3       BALB/cJ adult               M     <NA>         
 6     1    16  2019 Alk6       C57BL/6 embryo              <NA>  brain        
 7     1    19  2018 Alk3       BALB/cJ adult               F     <NA>         
 8     1    19  2019 Vegfr      C57BL/6 pup                 M     lung         
 9     1    20  2018 Vegfr      C57BL/6 pup                 F     lung         
10     1    24  2019 Vegfr      C57BL/6 adult               M     lung         
# … with 90 more rows, and 3 more variables: genotype <chr>,
#   experiment_type <chr>, researcher <chr>
~~~
{: .output}



~~~
samples[-(7:100), ] # equivalent to head(samples)
~~~
{: .language-r}



~~~
# A tibble: 6 x 12
  animal_id month   day  year mouse_line strain  developmental_stage sex  
      <dbl> <dbl> <dbl> <dbl> <chr>      <chr>   <chr>               <chr>
1    800793     1     6  2018 Alk3       BALB/cJ adult               F    
2    804396     1     7  2019 Vegfr      C57BL/6 adult               M    
3    805431     1    12  2018 Vegfr      C57BL/6 pup                 F    
4    805992     1    13  2019 Vegfr      C57BL/6 pup                 M    
5    808935     1    14  2020 Alk3       BALB/cJ adult               M    
6    810875     1    16  2019 Alk6       C57BL/6 embryo              <NA> 
# … with 4 more variables: organism_part <chr>, genotype <chr>,
#   experiment_type <chr>, researcher <chr>
~~~
{: .output}

Data frames can be subset by calling indices (as shown previously), but also by
calling their column names directly:


~~~
samples["animal_id"]
~~~
{: .language-r}



~~~
# A tibble: 100 x 1
   animal_id
       <dbl>
 1    800793
 2    804396
 3    805431
 4    805992
 5    808935
 6    810875
 7    812308
 8    814334
 9    816649
10    819947
# … with 90 more rows
~~~
{: .output}



~~~
samples[, "animal_id"]
~~~
{: .language-r}



~~~
# A tibble: 100 x 1
   animal_id
       <dbl>
 1    800793
 2    804396
 3    805431
 4    805992
 5    808935
 6    810875
 7    812308
 8    814334
 9    816649
10    819947
# … with 90 more rows
~~~
{: .output}

When we extract a subset from a data frame of class `tibble`, we normally
get back an object of the same class. To get one-column subsets returned as
vectors, we can use double square brackets:


~~~
samples[[1]]              # first column as vector
~~~
{: .language-r}



~~~
  [1] 800793 804396 805431 805992 808935 810875 812308 814334 816649 819947
 [11] 820421 821756 877817 821844 826176 832626 834217 835936 836507 837913
 [21] 842068 843132 844834 845290 855000 856721 862100 863932 864387 865043
 [31] 865149 866205 867481 867794 875004 876139 890389 886043 821080 889891
 [41] 890412 891135 891731 892354 893041 893257 893290 896197 898035 899372
 [51] 903535 904417 905633 909250 909687 910503 910988 916800 917874 918189
 [61] 918674 919295 920470 920574 924436 925791 928254 928944 932353 934662
 [71] 935401 935845 939519 941458 941646 942097 942925 943018 944091 945336
 [81] 946315 951520 951658 957895 958065 958342 959920 969452 971991 977235
 [91] 979356 980584 983907 985280 985570 986267 996829 996946 999154 999979
~~~
{: .output}



~~~
samples[["animal_id"]]    # named column as vector
~~~
{: .language-r}



~~~
  [1] 800793 804396 805431 805992 808935 810875 812308 814334 816649 819947
 [11] 820421 821756 877817 821844 826176 832626 834217 835936 836507 837913
 [21] 842068 843132 844834 845290 855000 856721 862100 863932 864387 865043
 [31] 865149 866205 867481 867794 875004 876139 890389 886043 821080 889891
 [41] 890412 891135 891731 892354 893041 893257 893290 896197 898035 899372
 [51] 903535 904417 905633 909250 909687 910503 910988 916800 917874 918189
 [61] 918674 919295 920470 920574 924436 925791 928254 928944 932353 934662
 [71] 935401 935845 939519 941458 941646 942097 942925 943018 944091 945336
 [81] 946315 951520 951658 957895 958065 958342 959920 969452 971991 977235
 [91] 979356 980584 983907 985280 985570 986267 996829 996946 999154 999979
~~~
{: .output}



~~~
samples[[1, 1]]           # first element in the first column as vector        
~~~
{: .language-r}



~~~
[1] 800793
~~~
{: .output}



~~~
samples[[1, "animal_id"]] # first element in the named column as vector
~~~
{: .language-r}



~~~
[1] 800793
~~~
{: .output}

We can also access an individual column as a vector by using a dollar sign, `$`:


~~~
samples$animal_id         # named column as vector
~~~
{: .language-r}



~~~
  [1] 800793 804396 805431 805992 808935 810875 812308 814334 816649 819947
 [11] 820421 821756 877817 821844 826176 832626 834217 835936 836507 837913
 [21] 842068 843132 844834 845290 855000 856721 862100 863932 864387 865043
 [31] 865149 866205 867481 867794 875004 876139 890389 886043 821080 889891
 [41] 890412 891135 891731 892354 893041 893257 893290 896197 898035 899372
 [51] 903535 904417 905633 909250 909687 910503 910988 916800 917874 918189
 [61] 918674 919295 920470 920574 924436 925791 928254 928944 932353 934662
 [71] 935401 935845 939519 941458 941646 942097 942925 943018 944091 945336
 [81] 946315 951520 951658 957895 958065 958342 959920 969452 971991 977235
 [91] 979356 980584 983907 985280 985570 986267 996829 996946 999154 999979
~~~
{: .output}

In RStudio, you can use the autocompletion feature to get the full and correct
names of the columns.

> ## Challenge
>
> 1. Create a `data.frame` (`samples_20`) containing only the data in
>    row 20 of the `samples` dataset.
>
> 2. Notice how `nrow()` gave you the number of rows in a `data.frame`?
>
>      * Use that number to pull out just that last row in the data frame.
>      * Compare that with what you see as the last row using `tail()` to make
>        sure it's meeting expectations.
>      * Pull out that last row using `nrow()` instead of the row number.
>      * Create a new data frame (`samples_last`) from that last row.
>
> 3. Use `nrow()` to extract the row that is in the middle of the data
>    frame. Store the content of this row in an object named `samples_middle`.
>
> 4. Combine `nrow()` with the `-` notation above to reproduce the behavior of
>    `head(samples)`, keeping just the first through 6th rows of the samples
>    dataset.
>
>> ## Solution
>>
>>
>>~~~
>>## 1.
>>samples_20 <- samples[20, ]
>>## 2.
>># Saving `n_rows` to improve readability and reduce duplication
>>n_rows <- nrow(samples)
>>samples_last <- samples[n_rows, ]
>>## 3.
>>samples_middle <- samples[n_rows / 2, ]
>>## 4.
>>samples_head<- samples[-(7:n_rows), ]
>>~~~
>>{: .language-r}
> {: .solution}
{: .challenge}


## Factors



When we did `str(samples)` we saw that the column `animal_id`, `month`, `day`
and `year` consist of numeric values. The remaining columns (e.g. `mouse_line`,
`strain` and `sex`), however, are of the class `character`. Arguably, these
columns contain categorical data, which means that they can only take on a
limited number of values. 

R has a special class for working with categorical data, called `factor`. 
Factors are very useful and actually contribute to making R particularly well 
suited to working with data. So we are going to spend a little time introducing 
them.

Once created, factors can only contain a pre-defined set of values, known as
*levels*. Factors are stored as integers associated with labels and they can be
ordered or unordered. While factors look (and often behave) like character
vectors, they are actually treated as integer vectors by R. So you need to be
very careful when treating them as strings.

When importing a data frame with `read_csv()`, the columns that contain text are
not automatically coerced (=converted) into the `factor` data type, but once we
have loaded the data we can do the conversion using the `factor()` function: 


~~~
samples$sex <- factor(samples$sex)
~~~
{: .language-r}

We can see that the conversion has worked by using the `summary()` 
function again. This produces a table with the counts for each factor level:


~~~
summary(samples$sex)
~~~
{: .language-r}



~~~
   F    M NA's 
  36   41   23 
~~~
{: .output}

By default, R always sorts levels in alphabetical order. For
instance, if you have a factor with 2 levels:


~~~
sex <- factor(c("male", "female", "female", "male"))
~~~
{: .language-r}

R will assign `1` to the level `"female"` and `2` to the level `"male"`
(because `f` comes before `m`, even though the first element in this vector is
`"male"`). You can see this by using the function `levels()` and you can find
the number of levels using `nlevels()`:


~~~
levels(sex)
~~~
{: .language-r}



~~~
[1] "female" "male"  
~~~
{: .output}



~~~
nlevels(sex)
~~~
{: .language-r}



~~~
[1] 2
~~~
{: .output}

Sometimes, the order of the factors does not matter, other times you might want
to specify the order because it is meaningful (e.g., "low", "medium", "high"),
it improves your visualization, or it is required by a particular type of
analysis. Here, one way to reorder our levels in the `sex` vector would be:


~~~
sex # current order
~~~
{: .language-r}



~~~
[1] male   female female male  
Levels: female male
~~~
{: .output}



~~~
sex <- factor(sex, levels = c("male", "female"))
sex # after re-ordering
~~~
{: .language-r}



~~~
[1] male   female female male  
Levels: male female
~~~
{: .output}

In R's memory, these factors are represented by integers (1, 2, 3), but are more
informative than integers because factors are self describing: `"female"`,
`"male"` is more descriptive than `1`, `2`. Which one is "male"?  You wouldn't
be able to tell just from the integer data. Factors, on the other hand, have
this information built in. It is particularly helpful when there are many levels
(like the species names in our example dataset).


> ## Challenge
>
> 1. Change the columns `strain` and `organism_part` in the `samples`
>    data frame into a factor.
>
> 2. Using the functions you have learnt so far, can you find out...
>
>      * How many samples were taken from brain tissue?
>      * How many different mouse strains are in the `strain` column?
>
>> ## Solution
>>
>> 
>> ~~~
>> samples$strain <- factor(samples$strain)
>> samples$organism_part <- factor(samples$organism_part)
>> summary(samples$organism_part)
>> ~~~
>> {: .language-r}
>> 
>> 
>> 
>> ~~~
>> brain  lung  NA's 
>>    16    58    26 
>> ~~~
>> {: .output}
>> 
>> 
>> 
>> ~~~
>> nlevels(samples$strain)
>> ~~~
>> {: .language-r}
>> 
>> 
>> 
>> ~~~
>> [1] 2
>> ~~~
>> {: .output}
>>
>> * How many brain samples? There are 16 "brain" values in the `organism_part`
     column.
>> * how many mouse strains? There are 2 strains in the `strain` column.
> {: .solution}
{: .challenge}


### Converting factors

If you need to convert a factor to a character vector, you use
`as.character(x)`.


~~~
as.character(sex)
~~~
{: .language-r}



~~~
[1] "male"   "female" "female" "male"  
~~~
{: .output}

In some cases, you may have to convert factors where the levels appear as
numbers (such as concentration levels or years) to a numeric vector. For
instance, in one part of your analysis the years might need to be encoded as
factors (e.g., comparing average weights across years) but in another part of
your analysis they may need to be stored as numeric values (e.g., doing math
operations on the years). This conversion from factor to numeric is a little
trickier. The `as.numeric()` function returns the index values of the factor,
not its levels, so it will result in an entirely new (and unwanted in this case)
set of numbers. One method to avoid this is to convert factors to characters,
and then to numbers.

Another method is to use the `levels()` function. Compare:


~~~
year_fct <- factor(c(1990, 1983, 1977, 1998, 1990))
as.numeric(year_fct)               # Wrong! And there is no warning...
~~~
{: .language-r}



~~~
[1] 3 2 1 4 3
~~~
{: .output}



~~~
as.numeric(as.character(year_fct)) # Works...
~~~
{: .language-r}



~~~
[1] 1990 1983 1977 1998 1990
~~~
{: .output}



~~~
as.numeric(levels(year_fct))[year_fct]    # The recommended way.
~~~
{: .language-r}



~~~
[1] 1990 1983 1977 1998 1990
~~~
{: .output}

Notice that in the `levels()` approach, three important steps occur:

* We obtain all the factor levels using `levels(year_fct)`
* We convert these levels to numeric values using `as.numeric(levels(year_fct))`
* We then access these numeric values using the underlying integers of the
  vector `year_fct` inside the square brackets

### Renaming factors

When your data is stored as a factor, you can use the `plot()` function to get a
quick glance at the number of observations represented by each factor
level. Let's look at the number of males and females captured over the course of
the experiment:


~~~
plot(samples$sex)
~~~
{: .language-r}

<img src="../fig/rmd-03-unnamed-chunk-25-1.png" title="plot of chunk unnamed-chunk-25" alt="plot of chunk unnamed-chunk-25" width="432" style="display: block; margin: auto;" />

However, as we saw when we used `summary(samples$sex)`, there are 23 individuals
for which the sex information hasn't been recorded. To show them in the plot, we
can turn the missing values into a factor level with the  `addNA()` function. We
will also have to give the new factor level a label. We are going to work with a
copy of the `sex` column, so we're not modifying the working copy of the data
frame:


~~~
sex <- samples$sex
levels(sex)
~~~
{: .language-r}



~~~
[1] "F" "M"
~~~
{: .output}



~~~
sex <- addNA(sex)
levels(sex)
~~~
{: .language-r}



~~~
[1] "F" "M" NA 
~~~
{: .output}



~~~
head(sex)
~~~
{: .language-r}



~~~
[1] F    M    F    M    M    <NA>
Levels: F M <NA>
~~~
{: .output}



~~~
levels(sex)[3] <- "unknown"
levels(sex)
~~~
{: .language-r}



~~~
[1] "F"       "M"       "unknown"
~~~
{: .output}



~~~
head(sex)
~~~
{: .language-r}



~~~
[1] F       M       F       M       M       unknown
Levels: F M unknown
~~~
{: .output}

Now we can plot the data again, using `plot(sex)`.

<img src="../fig/rmd-03-unnamed-chunk-27-1.png" title="plot of chunk unnamed-chunk-27" alt="plot of chunk unnamed-chunk-27" width="432" style="display: block; margin: auto;" />

> ## Challenge
>
> * Rename "F" and "M" to "female" and "male" respectively.
> * Now that we have renamed the factor level to "unknown", can you recreate
>   the barplot such that "unknown" is first (before "female")?
>
>> ## Solution
>>
>> 
>> ~~~
>> levels(sex)[1:2] <- c("female", "male")
>> sex <- factor(sex, levels = c("unknown", "female", "male"))
>> plot(sex)
>> ~~~
>> {: .language-r}
>> 
>> <img src="../fig/rmd-03-unnamed-chunk-28-1.png" title="plot of chunk unnamed-chunk-28" alt="plot of chunk unnamed-chunk-28" width="432" style="display: block; margin: auto;" />
> {: .solution}
{: .challenge}


> ## Challenge
>
> 1. We have seen how data frames are created when using `read_csv()`, but they
>   can also be created by hand with the `data.frame()` function.  There are a
>   few mistakes in this hand-crafted `data.frame`. Can you spot and fix them?
>   Don't hesitate to experiment!
>
>     
>     ~~~
>     animal_data <- data.frame(
>               animal = c(dog, cat, sea cucumber, sea urchin),
>               feel = c("furry", "squishy", "spiny"),
>               weight = c(45, 8 1.1, 0.8)
>               )
>     ~~~
>     {: .language-r}
>
> 2. Can you predict the class for each of the columns in the following example?
>    Check your guesses using `str(country_climate)`:
>      * Are they what you expected? Why? Why not?
>      * What would you need to change to ensure that each column had the
>        accurate data type?
>
>     
>     ~~~
>     country_climate <- data.frame(
>            country = c("Canada", "Panama", "South Africa", "Australia"),
>            climate = c("cold", "hot", "temperate", "hot/temperate"),
>            temperature = c(10, 30, 18, "15"),
>            northern_hemisphere = c(TRUE, TRUE, FALSE, "FALSE"),
>            has_kangaroo = c(FALSE, FALSE, FALSE, 1)
>            )
>     ~~~
>     {: .language-r}
>
>> ## Solution
>>
>> * missing quotations around the names of the animals
>> * missing one entry in the `feel` column (probably for one of the furry animals)
>> * missing one comma in the `weight` column
>> * `country`, `climate`, `temperature`, and `northern_hemisphere` are
>>    characters; `has_kangaroo` is numeric
>> * using `factor()` one could replace character columns with factors columns
>> * removing the quotes in `temperature` and `northern_hemisphere` and replacing 1
>>   by TRUE in the `has_kangaroo` column would give what was probably 
>>   intended
> {: .solution}
{: .challenge}

The automatic conversion of data type is sometimes a blessing, sometimes an
annoyance. Be aware that it exists, learn the rules, and double check that data
you import in R are of the correct type within your data frame. If not, use it
to your advantage to detect mistakes that might have been introduced during data
entry (for instance, a letter in a column that should only contain numbers).

Learn more in this [RStudio tutorial](https://support.rstudio.com/hc/en-us/articles/218611977-Importing-Data-with-RStudio)

## Formatting dates

One of the most common issues that new (and experienced!) R users have is
converting date and time information into a variable that is appropriate and
usable during analyses. As a reminder from earlier in this lesson, the best
practice for dealing with date data is to ensure that each component of your
date is stored as a separate variable. Using `str()`, We can confirm that our
data frame has a separate column for day, month, and year, and that each
contains integer values.


~~~
str(samples)
~~~
{: .language-r}
We are going to use the `ymd()` function from the package **`lubridate`**
(which belongs to the **`tidyverse`**; learn more [here](https://www.tidyverse.org/)).
**`lubridate`** gets installed as part of the **`tidyverse`** installation.
When you load the **`tidyverse`** (`library(tidyverse)`), the core packages
(the packages used in most data analyses) get loaded. **`lubridate`** however
does not belong to the core tidyverse, so you have to load it explicitly with
`library(lubridate)`

Start by loading the required package:


~~~
library(lubridate)
~~~
{: .language-r}

`ymd()` takes a vector representing year, month, and day, and converts it to a
`Date` vector. `Date` is a class of data recognized by R as being a date and can
be manipulated as such. The argument that the function requires is flexible,
but, as a best practice, is a character vector formatted as "YYYY-MM-DD".


Let's create a date object and inspect the structure:


~~~
my_date <- ymd("2015-01-01")
str(my_date)
~~~
{: .language-r}



~~~
 Date[1:1], format: "2015-01-01"
~~~
{: .output}

Now let's paste the year, month, and day separately - we get the same result:


~~~
# sep indicates the character to use to separate each component
my_date <- ymd(paste("2015", "1", "1", sep = "-")) 
str(my_date)
~~~
{: .language-r}



~~~
 Date[1:1], format: "2015-01-01"
~~~
{: .output}

Now we apply this function to the samples dataset. Create a character vector
from the `year`, `month`, and `day` columns of `samples` using `paste()`:


~~~
paste(samples$year, samples$month, samples$day, sep = "-")
~~~
{: .language-r}



~~~
  [1] "2018-1-6"   "2019-1-7"   "2018-1-12"  "2019-1-13"  "2020-1-14" 
  [6] "2019-1-16"  "2018-1-19"  "2019-1-19"  "2018-1-20"  "2019-1-24" 
 [11] "2019-1-31"  "2018-2-30"  "2019-4-13"  "2018-2-7"   "2019-2-14" 
 [16] "2020-2-16"  "2020-2-18"  "2020-2-18"  "2020-2-23"  "2020-2-27" 
 [21] "2020-2-27"  "2018-3-0"   "2018-3-7"   "2019-3-7"   "2019-3-9"  
 [26] "2020-3-9"   "2018-3-17"  "2018-3-21"  "2018-3-23"  "2019-3-29" 
 [31] "2019-3-30"  "2018-4-1"   "2018-4-10"  "2020-4-10"  "2019-4-11" 
 [36] "2018-4-13"  "2020-4-22"  "2019-4-17"  "2019-2-3"   "2018-4-21" 
 [41] "2020-4-22"  "2020-4-25"  "2020-4-30"  "2019-5-1"   "2019-5-3"  
 [46] "2020-5-3"   "2018-5-8"   "2018-5-9"   "2020-5-15"  "2018-5-18" 
 [51] "2019-5-25"  "2019-5-28"  "2018-6-4"   "2020-6-4"   "2020-6-6"  
 [56] "2019-6-7"   "2019-6-7"   "2020-6-7"   "2020-6-15"  "2020-6-22" 
 [61] "2020-6-28"  "2019-7-3"   "2019-7-4"   "2020-7-7"   "2018-7-12" 
 [66] "2018-7-15"  "2020-7-23"  "2019-7-25"  "2019-7-26"  "2018-7-27" 
 [71] "2019-8-3"   "2018-8-10"  "2020-8-10"  "2020-8-11"  "2020-8-23" 
 [76] "2020-8-24"  "2020-8-31"  "2019-9-2"   "2019-9-10"  "2018-9-15" 
 [81] "2018-9-16"  "2019-9-27"  "2019-10-3"  "2020-10-8"  "2019-10-11"
 [86] "2018-10-13" "2018-10-25" "2019-10-29" "2019-11-1"  "2019-11-2" 
 [91] "2019-11-3"  "2018-11-9"  "2019-11-14" "2019-11-21" "2018-12-18"
 [96] "2019-12-20" "2019-12-20" "2019-12-23" "2018-12-25" "2018-12-29"
~~~
{: .output}

This character vector can be used as the argument for `ymd()`:


~~~
ymd(paste(samples$year, samples$month, samples$day, sep = "-"))
~~~
{: .language-r}



~~~
Warning: 2 failed to parse.
~~~
{: .warning}



~~~
  [1] "2018-01-06" "2019-01-07" "2018-01-12" "2019-01-13" "2020-01-14"
  [6] "2019-01-16" "2018-01-19" "2019-01-19" "2018-01-20" "2019-01-24"
 [11] "2019-01-31" NA           "2019-04-13" "2018-02-07" "2019-02-14"
 [16] "2020-02-16" "2020-02-18" "2020-02-18" "2020-02-23" "2020-02-27"
 [21] "2020-02-27" NA           "2018-03-07" "2019-03-07" "2019-03-09"
 [26] "2020-03-09" "2018-03-17" "2018-03-21" "2018-03-23" "2019-03-29"
 [31] "2019-03-30" "2018-04-01" "2018-04-10" "2020-04-10" "2019-04-11"
 [36] "2018-04-13" "2020-04-22" "2019-04-17" "2019-02-03" "2018-04-21"
 [41] "2020-04-22" "2020-04-25" "2020-04-30" "2019-05-01" "2019-05-03"
 [46] "2020-05-03" "2018-05-08" "2018-05-09" "2020-05-15" "2018-05-18"
 [51] "2019-05-25" "2019-05-28" "2018-06-04" "2020-06-04" "2020-06-06"
 [56] "2019-06-07" "2019-06-07" "2020-06-07" "2020-06-15" "2020-06-22"
 [61] "2020-06-28" "2019-07-03" "2019-07-04" "2020-07-07" "2018-07-12"
 [66] "2018-07-15" "2020-07-23" "2019-07-25" "2019-07-26" "2018-07-27"
 [71] "2019-08-03" "2018-08-10" "2020-08-10" "2020-08-11" "2020-08-23"
 [76] "2020-08-24" "2020-08-31" "2019-09-02" "2019-09-10" "2018-09-15"
 [81] "2018-09-16" "2019-09-27" "2019-10-03" "2020-10-08" "2019-10-11"
 [86] "2018-10-13" "2018-10-25" "2019-10-29" "2019-11-01" "2019-11-02"
 [91] "2019-11-03" "2018-11-09" "2019-11-14" "2019-11-21" "2018-12-18"
 [96] "2019-12-20" "2019-12-20" "2019-12-23" "2018-12-25" "2018-12-29"
~~~
{: .output}

There is a warning telling us that some dates could not be parsed (understood)
by the `ymd()` function. For these dates, the function has returned `NA`, which
means they are treated as missing values.
We will deal with this problem later, but first we add the resulting `Date`
vector to the `samples` data frame as a new column called `date`:


~~~
samples$date <- ymd(paste(samples$year, samples$month, samples$day, sep = "-"))
~~~
{: .language-r}



~~~
Warning: 2 failed to parse.
~~~
{: .warning}



~~~
str(samples) # notice the new column, with 'date' as the class
~~~
{: .language-r}



~~~
spec_tbl_df[,13] [100 × 13] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ animal_id          : num [1:100] 800793 804396 805431 805992 808935 ...
 $ month              : num [1:100] 1 1 1 1 1 1 1 1 1 1 ...
 $ day                : num [1:100] 6 7 12 13 14 16 19 19 20 24 ...
 $ year               : num [1:100] 2018 2019 2018 2019 2020 ...
 $ mouse_line         : chr [1:100] "Alk3" "Vegfr" "Vegfr" "Vegfr" ...
 $ strain             : Factor w/ 2 levels "BALB/cJ","C57BL/6": 1 2 2 2 1 2 1 2 2 2 ...
 $ developmental_stage: chr [1:100] "adult" "adult" "pup" "pup" ...
 $ sex                : Factor w/ 2 levels "F","M": 1 2 1 2 2 NA 1 2 1 2 ...
 $ organism_part      : Factor w/ 2 levels "brain","lung": NA 2 2 2 NA 1 NA 2 2 2 ...
 $ genotype           : chr [1:100] "wild type" "wild type" "wild type" "wild type" ...
 $ experiment_type    : chr [1:100] "behaviour" "sequencing assay" "IHC" "IHC" ...
 $ researcher         : chr [1:100] "Kim" "Sam" "Sam" "Sam" ...
 $ date               : Date[1:100], format: "2018-01-06" "2019-01-07" ...
 - attr(*, "spec")=
  .. cols(
  ..   animal_id = col_double(),
  ..   month = col_double(),
  ..   day = col_double(),
  ..   year = col_double(),
  ..   mouse_line = col_character(),
  ..   strain = col_character(),
  ..   developmental_stage = col_character(),
  ..   sex = col_character(),
  ..   organism_part = col_character(),
  ..   genotype = col_character(),
  ..   experiment_type = col_character(),
  ..   researcher = col_character()
  .. )
~~~
{: .output}

Let's make sure everything worked correctly. One way to inspect the new column
is to use `summary()`:


~~~
summary(samples$date)
~~~
{: .language-r}



~~~
        Min.      1st Qu.       Median         Mean      3rd Qu.         Max. 
"2018-01-06" "2018-10-28" "2019-06-07" "2019-06-08" "2020-02-21" "2020-10-08" 
        NA's 
         "2" 
~~~
{: .output}

Let's investigate why some dates could not be parsed.

We can use the functions we saw previously to deal with missing data to identify
the rows in our data frame that are failing. If we combine them with what we
learned about subsetting data frames earlier, we can extract the columns "year,
"month", "day" from the records that have `NA` in our new column `date`. We will
also use `head()` so we don't clutter the output:


~~~
missing_dates <- samples[is.na(samples$date), c("year", "month", "day")]

head(missing_dates)
~~~
{: .language-r}



~~~
# A tibble: 2 x 3
   year month   day
  <dbl> <dbl> <dbl>
1  2018     2    30
2  2018     3     0
~~~
{: .output}

Why did these dates fail to parse? If you had to use these data for your
analyses, how would you deal with this situation?
