---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 04-dplyr.md in _episodes_rmd/
title: "Aggregating and Analyzing Data with dplyr"
teaching: 40
exercises: 15
questions:
- "How can I manipulate dataframes without repeating myself?"
objectives:
- "Describe the purpose of the **`dplyr`** and **`tidyr`** packages."
- "Select certain columns in a data frame with the **`dplyr`** function
  `select`."
- "Extract certain rows in a data frame according to logical (boolean)
  conditions with the **`dplyr`** function `filter` ."
- "Link the output of one **`dplyr`** function to the input of another function
  with the 'pipe' operator `%>%`."
- "Add new columns to a data frame that are functions of existing columns with
  `mutate`."
- "Use the split-apply-combine concept for data analysis."
- Use `summarize`, `group_by`, and `count` to split a data frame into groups of
  observations, apply summary statistics for each group, and then combine the
  results."
- "Describe the concept of a wide and a long table format and for which purpose
  those formats are useful."
- "Describe what key-value pairs are."
- "Reshape a data frame from long to wide format and back with the `spread` and
  `gather` commands from the **`tidyr`** package."
- "Export a data frame to a .csv file."
keypoints:
- "Use the `dplyr` package to manipulate dataframes."
- "Use `select()` to choose variables from a dataframe."
- "Use `filter()` to choose data based on values."
- "Use `group_by()` and `summarize()` to work with subsets of data."
- "Use `mutate()` to create new variables."
- "Use `pivot_wider()` to transform rows into columns."
- "Use `pivot_longer()` to transform columns into rows."
source: Rmd
---



# Data manipulation using **`dplyr`** and **`tidyr`**

Bracket subsetting is handy, but it can be cumbersome and difficult to read,
especially for complicated operations. Enter **`dplyr`**. **`dplyr`** is a
package for making tabular data manipulation easier. It pairs nicely with
**`tidyr`** which enables you to swiftly convert between different data formats
for plotting and analysis.

The **`tidyverse`** package is an "umbrella-package" that installs **`tidyr`**,
**`dplyr`**, and several other packages useful for data analysis, such as 
**`ggplot2`**, **`tibble`**, etc.

The **`tidyverse`** package tries to address 3 common issues that arise when
doing data analysis with some of the functions that come with R:

1. The results from a base R function sometimes depend on the type of data.
2. Using R expressions in a non standard way, which can be confusing for new
   learners.
3. Hidden arguments, having default operations that new learners are not aware
   of.

You should already have installed and loaded the **`tidyverse`** package. 
If we haven't already done so, we can type `install.packages("tidyverse")`
straight into the console. Then, to load the package type `library(tidyverse)`

## What are **`dplyr`** and **`tidyr`**?

The package **`dplyr`** provides easy tools for the most common data
manipulationvtasks. It is built to work directly with data frames, with many
common tasks optimized by being written in a compiled language (C++). An
additional feature is the ability to work directly with data stored in an
external database. The benefits of doing this are that the data can be managed
natively in a relational database, queries can be conducted on that database,
and only the results of the query are returned.

This addresses a common problem with R in that all operations are conducted
in-memory and thus the amount of data you can work with is limited by available
memory. The database connections essentially remove that limitation in that you
can connect to a database of many hundreds of GB, conduct queries on it
directly, and pull back into R only what you need for analysis.

The package **`tidyr`** addresses the common problem of wanting to reshape your
data for plotting and use by different R functions. Sometimes we want data sets
where we have one row per measurement. Sometimes we want a data frame where each
measurement type has its own column, and rows are instead more aggregated groups
(e.g., a time period, an experimental unit like a plot or a batch number).
Moving back and forth between these formats is non-trivial, and **`tidyr`**
gives you tools for this and more sophisticated  data manipulation.

To learn more about **`dplyr`** and **`tidyr`** after the workshop, you may want
to check out this [handy data transformation with **`dplyr`** cheatsheet](https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf)
and this [one about **`tidyr`**](https://github.com/rstudio/cheatsheets/raw/master/data-import.pdf).

In this episode, we will use data on gene expressions stored in the
[TSV (tab-separated values) format](https://en.wikipedia.org/wiki/Tab-separated_values).
We will read in our data using the `read_tsv()` function from the tidyverse
package **`readr`**.

Each line in the file holds the number of reads from are overlapping a
particular gene. Counts are given for 6 individual samples. The dataset consists
of the following columns:

| Column   | Description                            |
|----------|----------------------------------------|
| gene     | name of gene                           |
| sample   | 1-letter label for the sample (A to F) |
| genotype | "Wt" (wild type) or "Hom" (homozygous) |
| cnt      | number of reads overlapping the gene   |



~~~
download.file(
  url = "https://nbisweden.github.io/module-r-intro-dm-practices/data/gene_counts_r_lesson.csv",
  destfile = "data_raw/gene_counts_r_lesson.csv")
~~~
{: .language-r}


~~~
## load the tidyverse packages, incl. dplyr
library(tidyverse)
~~~
{: .language-r}

~~~
cnts <- read_tsv("data_raw/gene_counts_r_lesson.tsv")
~~~
{: .language-r}

~~~

── Column specification ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
cols(
  gene = col_character(),
  sample = col_character(),
  genotype = col_character(),
  cnt = col_double()
)
~~~
{: .output}


~~~
## inspect the data
str(cnts)
~~~
{: .language-r}



~~~
spec_tbl_df[,4] [259,680 × 4] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ gene    : chr [1:259680] "0610005C13Rik" "0610005C13Rik" "0610005C13Rik" "0610005C13Rik" ...
 $ sample  : chr [1:259680] "A" "B" "C" "D" ...
 $ genotype: chr [1:259680] "Wt" "Wt" "Wt" "Hom" ...
 $ cnt     : num [1:259680] 5 0 0 0 0 1 582 587 549 862 ...
 - attr(*, "spec")=
  .. cols(
  ..   gene = col_character(),
  ..   sample = col_character(),
  ..   genotype = col_character(),
  ..   cnt = col_double()
  .. )
~~~
{: .output}


~~~
## preview the data
view(cnts)
~~~
{: .language-r}

Next, we're going to learn some of the most common **`dplyr`** functions:

- `select()`: subset columns
- `filter()`: subset rows on conditions
- `pivot_wider()`: transform rows into columns
- `pivot_longer()`: transform columns into rows
- `mutate()`: create new columns by using information from other columns
- `group_by()` and `summarize()`: create summary statistics on grouped data
- `arrange()`: sort results
- `count()`: count discrete values

## Selecting columns and filtering rows

To select columns of a data frame, use `select()`. The first argument
to this function is the data frame (`cnts`), and the subsequent
arguments are the columns to keep.


~~~
select(cnts, gene, sample, cnt)
~~~
{: .language-r}

To select all columns *except* certain ones, put a "-" in front of
the variable to exclude it.


~~~
select(cnts, -sample, -genotype)
~~~
{: .language-r}

This will select all the variables in `cnts` except `sample`
and `genotype`.

To choose rows based on a specific criterion, use `filter()`:


~~~
filter(cnts, sample == "A")
~~~
{: .language-r}



~~~
# A tibble: 43,280 x 4
   gene          sample genotype   cnt
   <chr>         <chr>  <chr>    <dbl>
 1 0610005C13Rik A      Wt           5
 2 0610007P14Rik A      Wt         582
 3 0610009B22Rik A      Wt         132
 4 0610009E02Rik A      Wt          13
 5 0610009L18Rik A      Wt          36
 6 0610009O20Rik A      Wt         761
 7 0610010F05Rik A      Wt         201
 8 0610010K14Rik A      Wt           2
 9 0610011F06Rik A      Wt         307
10 0610012D04Rik A      Wt           4
# … with 43,270 more rows
~~~
{: .output}

## Pipes

What if you want to select and filter at the same time? There are three ways to
do this: use intermediate steps, nested functions, or pipes.

With intermediate steps, you create a temporary data frame and use that as input
to the next function, like this:


~~~
cnts_A <- filter(cnts, sample == "A")
cnts_A_sml <- select(cnts_A, gene, genotype, cnt)
~~~
{: .language-r}

This is readable, but can clutter up your workspace with lots of objects that
you have to name individually. With multiple steps, that can be hard to keep
track of.

You can also nest functions (i.e. one function inside of another), like this:


~~~
cnts_A_sml <- select(
  filter(cnts, sample == "A"), gene, genotype, cnt)
~~~
{: .language-r}

This is handy, but can be difficult to read if too many functions are nested, as
R evaluates the expression from the inside out (in this case, filtering, then
selecting).

The last option, *pipes*, are a recent addition to R. Pipes let you take the
output of one function and send it directly to the next, which is useful when
you need to do many things to the same dataset.  Pipes in R look like `%>%` and
are made available via the **`magrittr`** package, installed automatically with
**`dplyr`**. If you use RStudio, you can type the pipe with <kbd>Ctrl</kbd> +
<kbd>Shift</kbd> + <kbd>M</kbd> if you have a PC or <kbd>Cmd</kbd> +
<kbd>Shift</kbd> + <kbd>M</kbd> if you have a Mac.


~~~
cnts %>%
  filter(sample == "A") %>%
  select(gene, genotype, cnt)
~~~
{: .language-r}



~~~
# A tibble: 43,280 x 3
   gene          genotype   cnt
   <chr>         <chr>    <dbl>
 1 0610005C13Rik Wt           5
 2 0610007P14Rik Wt         582
 3 0610009B22Rik Wt         132
 4 0610009E02Rik Wt          13
 5 0610009L18Rik Wt          36
 6 0610009O20Rik Wt         761
 7 0610010F05Rik Wt         201
 8 0610010K14Rik Wt           2
 9 0610011F06Rik Wt         307
10 0610012D04Rik Wt           4
# … with 43,270 more rows
~~~
{: .output}

In the above code, we use the pipe to send the `cnts` dataset first
through `filter()` to keep rows where `sample` equals `"A"`, then through
`select()` to keep only the `gene`, `genotype`, and `cnt` columns. Since
`%>%` takes the object on its left and passes it as the first argument to the
function on its right, we don't need to explicitly include the data frame as an
argument to the `filter()` and `select()` functions any more.

Some may find it helpful to read the pipe like the word "then". For instance,
in the above example, we took the data frame `cnts`, *then* we `filter`ed
for rows with `sample == "A"`, *then* we `select`ed columns `gene`, `genotype`,
and `cnt`. The **`dplyr`** functions by themselves are somewhat simple,
but by combining them into linear workflows with the pipe, we can accomplish
more complex manipulations of data frames.

If we want to create a new object with this smaller version of the data, we
can assign it a new name:


~~~
cnts_A_sml <- cnts %>%
  filter(sample == "A") %>%
  select(gene, genotype, cnt)

cnts_A_sml
~~~
{: .language-r}



~~~
# A tibble: 43,280 x 3
   gene          genotype   cnt
   <chr>         <chr>    <dbl>
 1 0610005C13Rik Wt           5
 2 0610007P14Rik Wt         582
 3 0610009B22Rik Wt         132
 4 0610009E02Rik Wt          13
 5 0610009L18Rik Wt          36
 6 0610009O20Rik Wt         761
 7 0610010F05Rik Wt         201
 8 0610010K14Rik Wt           2
 9 0610011F06Rik Wt         307
10 0610012D04Rik Wt           4
# … with 43,270 more rows
~~~
{: .output}

Note that the final data frame is the leftmost part of this expression.

> ## Challenge 4.1
>
> Using pipes, subset the `cnts` data to include only counts for wild type
> genotypes above 10,000 and retain only the columns `gene` and `cnt`.
>
>> ## Solution
>>
>> 
>> ~~~
>> cnts %>%
>>   filter(cnt > 10000 & genotype == "Wt") %>%
>>   select(gene, cnt)
>> ~~~
>> {: .language-r}
>> 
>> 
>> 
>> ~~~
>> # A tibble: 399 x 2
>>    gene     cnt
>>    <chr>  <dbl>
>>  1 Abca3  10111
>>  2 Ace    13834
>>  3 Ace    16353
>>  4 Ace    16830
>>  5 Actb   68901
>>  6 Actb   62553
>>  7 Actb   68037
>>  8 Acvrl1 10195
>>  9 Ager   51453
>> 10 Ager   47068
>> # … with 389 more rows
>> ~~~
>> {: .output}
> {: .solution}
{: .challenge}


### Reshaping with pivot_wider and pivot_longer

You may need to alter the shape of a dataset before you can analyze it.
A general recommendation for tidy data is that you should put each observation 
in its own row, and each variable (i.e. the thing you are measuring) in a
separate column. What you count as an observation depends on the question you
are addressing, and you need to reshape your data accordingly. You can read
more about the principles behind tidy data in Data Carpentry's
[spreadsheet lesson](https://datacarpentry.org/spreadsheet-ecology-lesson/01-format-data/).

There are basically two ways to reshape data frames. One is to turn rows into
columns by transforming values of a variable to column names. The other is to 
turn columns into rows by transforming column names into values of a variable.
We can do these of transformations with the functions `pivot_wider()` and
`pivot_longer()`, respectively.

#### pivot_wider

`pivot_wider()` takes four principal arguments:

1. the data
2. the *id_cols* with the set of variables that uniquely identifies each
   observation.
3. the *names_from* variable to get the output column names from
4. the *values_from* column variable whose values will fill the new column
   variables.


Let's use `pivot_wider()` to transform the gene count data so that we get all
the 6 counts for one gene in the same row.


~~~
cnts %>% pivot_wider(
  id_cols = gene,
  names_from = sample,
  values_from = cnt
)
~~~
{: .language-r}



~~~
# A tibble: 43,280 x 7
   gene              A     B     C     D     E     F
   <chr>         <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
 1 0610005C13Rik     5     0     0     0     0     1
 2 0610007P14Rik   582   587   549   862   755   836
 3 0610009B22Rik   132    95   119   197   233   200
 4 0610009E02Rik    13     6     9     7    18    17
 5 0610009L18Rik    36    25    32    39    41    58
 6 0610009O20Rik   761   855   792  1295  1352  1233
 7 0610010F05Rik   201   186   171   330   408   309
 8 0610010K14Rik     2     1    11     3     7     6
 9 0610011F06Rik   307   317   318   377   357   434
10 0610012D04Rik     4     0     2     3     4     0
# … with 43,270 more rows
~~~
{: .output}

This looks pretty good, but we are missing the `genotype` variable. Let's
add this information to the column names as well.


~~~
cnts_wide <- cnts %>%
  pivot_wider(
    id_cols = gene,
    names_from = c(sample, genotype),
    values_from = cnt
)

str(cnts_wide)
~~~
{: .language-r}



~~~
tibble[,7] [43,280 × 7] (S3: tbl_df/tbl/data.frame)
 $ gene : chr [1:43280] "0610005C13Rik" "0610007P14Rik" "0610009B22Rik" "0610009E02Rik" ...
 $ A_Wt : num [1:43280] 5 582 132 13 36 761 201 2 307 4 ...
 $ B_Wt : num [1:43280] 0 587 95 6 25 855 186 1 317 0 ...
 $ C_Wt : num [1:43280] 0 549 119 9 32 792 171 11 318 2 ...
 $ D_Hom: num [1:43280] 0 862 197 7 39 ...
 $ E_Hom: num [1:43280] 0 755 233 18 41 ...
 $ F_Hom: num [1:43280] 1 836 200 17 58 ...
~~~
{: .output}

We store the data frame under the name `cnts_wide` so that we can use it
later in this episode.


#### pivot_longer

The opposing situation could occur if we had been provided with data in the
form of `cnts_wide`, where the sample labels are column names, but we 
wish to treat them as values of a sample variable instead.

In this situation we are gathering the column names and turning them into a
pair of new variables. One variable represents the column names as values, and
the other variable contains the values previously associated with the column
names.

`pivot_longer()` takes four principal arguments:

1. the data
2. the *cols* with the set of columns we want transform into a variable.
3. the *names_to* column variable we wish to create and fill with values 
associated with the key.
4. the *values_to* column variable we want to fill with the values.


~~~
cnts_wide %>% pivot_longer(
    cols = A_Wt:F_Hom,
    names_to = c("sample", "genotype"),
    names_sep = "_",
    values_to = "cnt"
)
~~~
{: .language-r}



~~~
# A tibble: 259,680 x 4
   gene          sample genotype   cnt
   <chr>         <chr>  <chr>    <dbl>
 1 0610005C13Rik A      Wt           5
 2 0610005C13Rik B      Wt           0
 3 0610005C13Rik C      Wt           0
 4 0610005C13Rik D      Hom          0
 5 0610005C13Rik E      Hom          0
 6 0610005C13Rik F      Hom          1
 7 0610007P14Rik A      Wt         582
 8 0610007P14Rik B      Wt         587
 9 0610007P14Rik C      Wt         549
10 0610007P14Rik D      Hom        862
# … with 259,670 more rows
~~~
{: .output}

Since we are transforming column names into two new variables, we must specify a
string separator using the `names_sep` argument.


> ## Challenge 4.2
>
> Reshape the `cnts` data frame with samples as rows and the three genes
> "Hopx", "Rpl21" and "Calr" as columns.
>
>> ## Solution
>>
>> 
>> ~~~
>> cnts %>%
>>   filter(gene %in% c("Hopx", "Rpl21", "Calr")) %>%
>>   pivot_wider(
>>     id_cols = sample,
>>     names_from = gene,
>>     values_from = cnt
>>   )
>> ~~~
>> {: .language-r}
>> 
>> 
>> 
>> ~~~
>> # A tibble: 6 x 4
>>   sample  Calr  Hopx Rpl21
>>   <chr>  <dbl> <dbl> <dbl>
>> 1 A       6137 18765 12499
>> 2 B       6279 15569 11248
>> 3 C       6733 14070  9851
>> 4 D      12369 16010 14494
>> 5 E      10524 18048 14177
>> 6 F      10801 19600 15642
>> ~~~
>> {: .output}
> {: .solution}
{: .challenge}


### Mutate

Frequently you'll want to create new columns based on the values in existing
columns, for example to do unit conversions, or to find the ratio of values in
two columns. For this we'll use `mutate()`.

Let's start by calculating the difference between `D_Hom` and `E_Hom` variables
in the `cnts_wide` data frame:


~~~
cnts_wide %>%
  mutate(diff_DE = D_Hom - E_Hom)
~~~
{: .language-r}



~~~
# A tibble: 43,280 x 8
   gene           A_Wt  B_Wt  C_Wt D_Hom E_Hom F_Hom diff_DE
   <chr>         <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>   <dbl>
 1 0610005C13Rik     5     0     0     0     0     1       0
 2 0610007P14Rik   582   587   549   862   755   836     107
 3 0610009B22Rik   132    95   119   197   233   200     -36
 4 0610009E02Rik    13     6     9     7    18    17     -11
 5 0610009L18Rik    36    25    32    39    41    58      -2
 6 0610009O20Rik   761   855   792  1295  1352  1233     -57
 7 0610010F05Rik   201   186   171   330   408   309     -78
 8 0610010K14Rik     2     1    11     3     7     6      -4
 9 0610011F06Rik   307   317   318   377   357   434      20
10 0610012D04Rik     4     0     2     3     4     0      -1
# … with 43,270 more rows
~~~
{: .output}

You can also create a second new column based on the first new column within the
same call of `mutate()`:


~~~
cnts_wide %>%
  mutate(diff_DE = D_Hom - E_Hom,
         diff_DE_pct = diff_DE / D_Hom * 100)
~~~
{: .language-r}



~~~
# A tibble: 43,280 x 9
   gene           A_Wt  B_Wt  C_Wt D_Hom E_Hom F_Hom diff_DE diff_DE_pct
   <chr>         <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>   <dbl>       <dbl>
 1 0610005C13Rik     5     0     0     0     0     1       0      NaN   
 2 0610007P14Rik   582   587   549   862   755   836     107       12.4 
 3 0610009B22Rik   132    95   119   197   233   200     -36      -18.3 
 4 0610009E02Rik    13     6     9     7    18    17     -11     -157.  
 5 0610009L18Rik    36    25    32    39    41    58      -2       -5.13
 6 0610009O20Rik   761   855   792  1295  1352  1233     -57       -4.40
 7 0610010F05Rik   201   186   171   330   408   309     -78      -23.6 
 8 0610010K14Rik     2     1    11     3     7     6      -4     -133.  
 9 0610011F06Rik   307   317   318   377   357   434      20        5.31
10 0610012D04Rik     4     0     2     3     4     0      -1      -33.3 
# … with 43,270 more rows
~~~
{: .output}

If this runs off your screen and you just want to see the first few rows, you
can use a pipe to view the `head()` of the data. (Pipes work with 
non-**`dplyr`** functions, too, as long as the **`dplyr`** or `magrittr` package
is loaded).


~~~
cnts_wide %>%
  mutate(diff_DE = D_Hom - E_Hom,
         diff_DE_pct = diff_DE / D_Hom * 100) %>%
  head()
~~~
{: .language-r}



~~~
# A tibble: 6 x 9
  gene           A_Wt  B_Wt  C_Wt D_Hom E_Hom F_Hom diff_DE diff_DE_pct
  <chr>         <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>   <dbl>       <dbl>
1 0610005C13Rik     5     0     0     0     0     1       0      NaN   
2 0610007P14Rik   582   587   549   862   755   836     107       12.4 
3 0610009B22Rik   132    95   119   197   233   200     -36      -18.3 
4 0610009E02Rik    13     6     9     7    18    17     -11     -157.  
5 0610009L18Rik    36    25    32    39    41    58      -2       -5.13
6 0610009O20Rik   761   855   792  1295  1352  1233     -57       -4.40
~~~
{: .output}

`NaN`s ("Not a Number") will be inserted in the last column whenever we try to
divide a value by zero. If we want to remove those, we can insert a `filter()`
in the chain:


~~~
cnts_wide %>%
  mutate(diff_DE = D_Hom - E_Hom,
         diff_DE_pct = diff_DE / D_Hom * 100) %>%
  filter(!is.na(diff_DE_pct)) %>%
  head()
~~~
{: .language-r}



~~~
# A tibble: 6 x 9
  gene           A_Wt  B_Wt  C_Wt D_Hom E_Hom F_Hom diff_DE diff_DE_pct
  <chr>         <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>   <dbl>       <dbl>
1 0610007P14Rik   582   587   549   862   755   836     107       12.4 
2 0610009B22Rik   132    95   119   197   233   200     -36      -18.3 
3 0610009E02Rik    13     6     9     7    18    17     -11     -157.  
4 0610009L18Rik    36    25    32    39    41    58      -2       -5.13
5 0610009O20Rik   761   855   792  1295  1352  1233     -57       -4.40
6 0610010F05Rik   201   186   171   330   408   309     -78      -23.6 
~~~
{: .output}

`is.na()` is a function that determines whether something is an `NA` (or `NaN`).
The `!` symbol negates the result, so we're asking for every row where weight
*is not* an `NA` (or `NaN`).

> ## Challenge 4.3
>
> Create a new data frame from the `cnts_wide` data that meets the
> following criteria: contains only the `gene` column and a new column called
> `log_A_Wt` with the common logarithm of the `cnt` values in the `A_Wt` column.
>
> **Hint**: You can use the `log10()` function to calculate the common
> logarithm values.
> 
>> ## Solution
>>
>> 
>> ~~~
>> cnts_wide %>%
>>   mutate(log_A_Wt = log10(A_Wt)) %>%
>>   select(gene, log_A_Wt)
>> ~~~
>> {: .language-r}
>> 
>> 
>> 
>> ~~~
>> # A tibble: 43,280 x 2
>>    gene          log_A_Wt
>>    <chr>            <dbl>
>>  1 0610005C13Rik    0.699
>>  2 0610007P14Rik    2.76 
>>  3 0610009B22Rik    2.12 
>>  4 0610009E02Rik    1.11 
>>  5 0610009L18Rik    1.56 
>>  6 0610009O20Rik    2.88 
>>  7 0610010F05Rik    2.30 
>>  8 0610010K14Rik    0.301
>>  9 0610011F06Rik    2.49 
>> 10 0610012D04Rik    0.602
>> # … with 43,270 more rows
>> ~~~
>> {: .output}
> {: .solution}
{: .challenge}


### Split-apply-combine data analysis and the `summarize()` function

Many data analysis tasks can be approached using the *split-apply-combine*
paradigm: split the data into groups, apply some analysis to each group, and
then combine the results. **`dplyr`** makes this very easy through the use of
the `group_by()` function.


#### The `summarize()` function

`group_by()` is often used together with `summarize()`, which collapses each
group into a single-row summary of that group. `group_by()` takes as arguments
the column names that contain the **categorical** variables for which you want
to calculate the summary statistics. So to compute the mean `weight` by sex:


~~~
cnts %>%
  group_by(genotype) %>%
  summarize(mean = mean(cnt))
~~~
{: .language-r}



~~~
# A tibble: 2 x 2
  genotype  mean
  <chr>    <dbl>
1 Hom       507.
2 Wt        344.
~~~
{: .output}

You can also group by multiple columns:


~~~
cnts %>%
  group_by(gene, genotype) %>%
  summarize(mean = mean(cnt)) %>% 
  head()
~~~
{: .language-r}



~~~
# A tibble: 6 x 3
# Groups:   gene [3]
  gene          genotype    mean
  <chr>         <chr>      <dbl>
1 0610005C13Rik Hom        0.333
2 0610005C13Rik Wt         1.67 
3 0610007P14Rik Hom      818.   
4 0610007P14Rik Wt       573.   
5 0610009B22Rik Hom      210    
6 0610009B22Rik Wt       115.   
~~~
{: .output}

Once the data are grouped, you can also summarize multiple variables at the same
time (and not necessarily on the same variable). For instance, we could add a
column indicating the minimum count for each genotype for each gene:


~~~
cnts %>%
  group_by(gene, genotype) %>%
  summarize(mean = mean(cnt),
            min = min(cnt)) %>%
  head()
~~~
{: .language-r}



~~~
# A tibble: 6 x 4
# Groups:   gene [3]
  gene          genotype    mean   min
  <chr>         <chr>      <dbl> <dbl>
1 0610005C13Rik Hom        0.333     0
2 0610005C13Rik Wt         1.67      0
3 0610007P14Rik Hom      818.      755
4 0610007P14Rik Wt       573.      549
5 0610009B22Rik Hom      210       197
6 0610009B22Rik Wt       115.       95
~~~
{: .output}

It is sometimes useful to rearrange the result of a query to inspect the values.
For instance, we can sort on `min_cnt` to put the lowest numbers first:



~~~
cnts %>%
  group_by(gene, genotype) %>%
  summarize(mean = mean(cnt),
            min = min(cnt)) %>%
  arrange(min)
~~~
{: .language-r}



~~~
# A tibble: 86,560 x 4
# Groups:   gene [43,280]
   gene          genotype  mean   min
   <chr>         <chr>    <dbl> <dbl>
 1 0610005C13Rik Hom      0.333     0
 2 0610005C13Rik Wt       1.67      0
 3 0610012D04Rik Hom      2.33      0
 4 0610012D04Rik Wt       2         0
 5 0610025J13Rik Wt       1.33      0
 6 0610031O16Rik Hom      0.333     0
 7 0610031O16Rik Wt       0.333     0
 8 0610040F04Rik Hom      1.33      0
 9 1110002J07Rik Wt       1.67      0
10 1110008E08Rik Hom      0         0
# … with 86,550 more rows
~~~
{: .output}

To sort in descending order, we need to add the `desc()` function. If we want
to sort the results by decreasing order of mean count:


~~~
cnts %>%
  group_by(gene, genotype) %>%
  summarize(mean = mean(cnt),
            min = min(cnt)) %>%
  arrange(desc(min))
~~~
{: .language-r}



~~~
# A tibble: 86,560 x 4
# Groups:   gene [43,280]
   gene    genotype    mean    min
   <chr>   <chr>      <dbl>  <dbl>
 1 Yam1    Hom      562743. 546656
 2 Scgb1a1 Hom      343686. 312982
 3 Scgb1a1 Wt       405803. 312507
 4 Yam1    Wt       391803. 299211
 5 Sftpc   Hom      289526. 261788
 6 Sftpc   Wt       232192. 213059
 7 mt-Rnr2 Wt       173459. 159714
 8 Lyz2    Hom      152748  142628
 9 mt-Rnr2 Hom      153111. 141593
10 Sparc   Hom      108928  103711
# … with 86,550 more rows
~~~
{: .output}

#### Counting

When working with data, we often want to know the number of observations found
for each factor or combination of factors. For this task, **`dplyr`** provides
`count()`. For example, if we wanted to count the number of rows of data for
each sample, we would do:


~~~
cnts %>%
    count(sample) 
~~~
{: .language-r}



~~~
# A tibble: 6 x 2
  sample     n
  <chr>  <int>
1 A      43280
2 B      43280
3 C      43280
4 D      43280
5 E      43280
6 F      43280
~~~
{: .output}

The `count()` function is shorthand for something we've already seen: grouping
by a variable, and summarizing it by counting the number of observations in that
group. In other words, `cnts %>% count(sample)` is equivalent to:  


~~~
cnts %>%
    group_by(sample) %>%
    summarize(n = n())
~~~
{: .language-r}



~~~
# A tibble: 6 x 2
  sample     n
  <chr>  <int>
1 A      43280
2 B      43280
3 C      43280
4 D      43280
5 E      43280
6 F      43280
~~~
{: .output}

We can also combine `count()` with other functions such as `filter()`. Here
we will count for each sample the number of genes with more than 10,000 reads.


~~~
cnts %>%
    filter(cnt > 10000) %>%
    count(gene) 
~~~
{: .language-r}



~~~
# A tibble: 260 x 2
   gene              n
   <chr>         <int>
 1 9430020K01Rik     1
 2 Abca3             4
 3 Abhd2             1
 4 Ace               6
 5 Actb              6
 6 Acvrl1            4
 7 Aes               1
 8 Ager              6
 9 Ahnak             6
10 Akap5             2
# … with 250 more rows
~~~
{: .output}

The example above shows the use of `count()` to count the number of
rows/observations for *one* factor (i.e., `gene`). If we wanted to count
*combination of factors*, such as `gene` and `genotype`, we would specify the
first and the second factor as the arguments of `count()`:


~~~
cnts %>%
  filter(cnt > 10000) %>%
  count(gene, genotype) 
~~~
{: .language-r}



~~~
# A tibble: 419 x 3
   gene          genotype     n
   <chr>         <chr>    <int>
 1 9430020K01Rik Hom          1
 2 Abca3         Hom          3
 3 Abca3         Wt           1
 4 Abhd2         Hom          1
 5 Ace           Hom          3
 6 Ace           Wt           3
 7 Actb          Hom          3
 8 Actb          Wt           3
 9 Acvrl1        Hom          3
10 Acvrl1        Wt           1
# … with 409 more rows
~~~
{: .output}

With the above code, we can proceed with `arrange()` to sort the table 
according to a number of criteria so that we have a better comparison. 
For instance, we might want to arrange the table above in (i) an alphabetical
order of the levels of the genotypes and (ii) in descending order of the count:


~~~
cnts %>%
  filter(cnt > 10000) %>%
  count(gene, genotype) %>%
  arrange(genotype, desc(n))
~~~
{: .language-r}



~~~
# A tibble: 419 x 3
   gene    genotype     n
   <chr>   <chr>    <int>
 1 Abca3   Hom          3
 2 Ace     Hom          3
 3 Actb    Hom          3
 4 Acvrl1  Hom          3
 5 Ager    Hom          3
 6 Ahnak   Hom          3
 7 Akr1a1  Hom          3
 8 Aldh1a1 Hom          3
 9 Aldh2   Hom          3
10 Anxa2   Hom          3
# … with 409 more rows
~~~
{: .output}

From the table above, we may learn that, for instance, there are 75 observations of 
the *albigula* species that are not specified for its sex (i.e. `NA`).


> ## Challenge 4.4
>
> * For each sample in the `cnts` data frame, how many of the genes have
>   zero reads?
>
>> ## Solution
>>
>> 
>> ~~~
>> cnts %>%
>>   filter(cnt == 0) %>%
>>   count(gene) 
>> ~~~
>> {: .language-r}
>> 
>> 
>> 
>> ~~~
>> # A tibble: 23,451 x 2
>>    gene              n
>>    <chr>         <int>
>>  1 0610005C13Rik     4
>>  2 0610012D04Rik     2
>>  3 0610025J13Rik     1
>>  4 0610031O16Rik     4
>>  5 0610040F04Rik     2
>>  6 1110002J07Rik     1
>>  7 1110008E08Rik     6
>>  8 1110015O18Rik     6
>>  9 1110025L11Rik     6
>> 10 1110028F18Rik     5
>> # … with 23,441 more rows
>> ~~~
>> {: .output}
> {: .solution}
>
> * Use `group_by()` and `summarize()` to find the mean and standard deviation
>   of the number of reads for each genotype and gene. Then reshape the data
>   so that the mean and standard deviation for for the two genotypes are
>   displayed in separate columns.
>
>   **Hint:** calculate the standard deviation with the `sd()` function.
>
>> ## Solution
>>
>> 
>> ~~~
>> cnts %>%
>>     group_by(gene, genotype) %>%
>>     summarize(
>>         mean = mean(cnt),
>>         stdev = sd(cnt)
>>     ) %>% 
>>     pivot_wider(
>>         id_cols = gene,
>>         names_from = genotype,
>>         values_from = (c(mean, stdev))
>>     )
>> ~~~
>> {: .language-r}
>> 
>> 
>> 
>> ~~~
>> # A tibble: 43,280 x 5
>> # Groups:   gene [43,280]
>>    gene          mean_Hom mean_Wt stdev_Hom stdev_Wt
>>    <chr>            <dbl>   <dbl>     <dbl>    <dbl>
>>  1 0610005C13Rik    0.333    1.67     0.577     2.89
>>  2 0610007P14Rik  818.     573.      55.8      20.6 
>>  3 0610009B22Rik  210      115.      20.0      18.8 
>>  4 0610009E02Rik   14        9.33     6.08      3.51
>>  5 0610009L18Rik   46       31       10.4       5.57
>>  6 0610009O20Rik 1293.     803.      59.5      47.9 
>>  7 0610010F05Rik  349      186       52.2      15   
>>  8 0610010K14Rik    5.33     4.67     2.08      5.51
>>  9 0610011F06Rik  389.     314       40.0       6.08
>> 10 0610012D04Rik    2.33     2        2.08      2   
>> # … with 43,270 more rows
>> ~~~
>> {: .output}
> {: .solution}
{: .challenge}


# Exporting data

Now that you have learned how to use **`dplyr`** to extract information from
or summarize your raw data, you may want to export these new data sets to share
them with your collaborators or for archival.

Similar to the `read_csv()` function used for reading CSV files into R, there is
a `write_csv()` function that generates CSV files from data frames.

Before using `write_csv()`, we are going to create a new folder, `data`,
in our working directory that will store this generated dataset. We don't want
to write generated datasets in the same directory as our raw data. It's good
practice to keep them separate. The `data_raw` folder should only contain the
raw, unaltered data, and should be left alone to make sure we don't delete or
modify it. In contrast, our script will generate the contents of the `data`
directory, so even if the files it contains are deleted, we can always
re-generate them.

We will conclude this episode by generating two CSV files with a small dataset that
we will use in the next episode on plotting:


~~~
# Create a vector with a subset of genes
genes_subset <- c(
  "Rbpj", "Nop10", "Nsun2", "Cxcl5", "Cfdp1", "Furin", "Adcy8", "Grin1",
  "Itpka", "Ippk", "Aff2", "Rab43", "Ofd1", "Edem2", "Xlr4c", "Mpc2",
  "Ect2l", "Pcgf1", "Nov", "Sys1", "Ackr2", "Acat1", "Phkg2", "Hecw2",
  "Ctps", "Arl5b", "Wnt3a", "Evc", "Wibg", "Axl", "Usp19", "Agbl2",
  "Eif3f", "Mark4", "Sbk2", "Eri3", "Tab1", "Cd276", "Sox9", "Ces2e",
  "Htra3", "Foxj2", "Arl4d", "Wdr1", "Taco1", "Aebp2", "Nemf", "Fpgt",
  "Helq", "Nudt6")

# Filter out subset of genes
cnts_sml_wide <- cnts_wide %>%
  filter(gene %in% genes_subset)

cnts_sml_long <- cnts %>%
  filter(gene %in% genes_subset)

# Write data frame to CSV
write_csv(cnts_sml_wide, file = "gene_counts_sml_wide.csv")

# Write data frame to CSV
write_csv(cnts_sml_long, file = "gene_counts_sml_long.csv")
~~~
{: .language-r}
