---
title: "Aggregating and Analyzing Data with dplyr"
teaching: 40
exercises: 15
questions:
- "How can I manipulate dataframes without repeating myself?"
objectives:
- "Describe the purpose of the **`dplyr`** and **`tidyr`** packages."
- "Select certain columns in a data frame with the **`dplyr`** function
  `select`."
- "Extract certain rows in a data frame according to logical (boolean)
  conditions with the **`dplyr`** function `filter` ."
- "Link the output of one **`dplyr`** function to the input of another function
  with the 'pipe' operator `%>%`."
- "Add new columns to a data frame that are functions of existing columns with
  `mutate`."
- "Use the split-apply-combine concept for data analysis."
- Use `summarize`, `group_by`, and `count` to split a data frame into groups of
  observations, apply summary statistics for each group, and then combine the
  results."
- "Describe the concept of a wide and a long table format and for which purpose
  those formats are useful."
- "Describe what key-value pairs are."
- "Reshape a data frame from long to wide format and back with the `spread` and
  `gather` commands from the **`tidyr`** package."
- "Export a data frame to a .csv file."
keypoints:
- "Use the `dplyr` package to manipulate dataframes."
- "Use `select()` to choose variables from a dataframe."
- "Use `filter()` to choose data based on values."
- "Use `group_by()` and `summarize()` to work with subsets of data."
- "Use `mutate()` to create new variables."
source: Rmd
---

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("04-")
```

# Data manipulation using **`dplyr`** and **`tidyr`**

Bracket subsetting is handy, but it can be cumbersome and difficult to read,
especially for complicated operations. Enter **`dplyr`**. **`dplyr`** is a
package for making tabular data manipulation easier. It pairs nicely with
**`tidyr`** which enables you to swiftly convert between different data formats
for plotting and analysis.

The **`tidyverse`** package is an "umbrella-package" that installs **`tidyr`**,
**`dplyr`**, and several other packages useful for data analysis, such as 
**`ggplot2`**, **`tibble`**, etc.

The **`tidyverse`** package tries to address 3 common issues that arise when
doing data analysis with some of the functions that come with R:

1. The results from a base R function sometimes depend on the type of data.
2. Using R expressions in a non standard way, which can be confusing for new
   learners.
3. Hidden arguments, having default operations that new learners are not aware
   of.

You should already have installed and loaded the **`tidyverse`** package. 
If we haven't already done so, we can type `install.packages("tidyverse")`
straight into the console. Then, to load the package type `library(tidyverse)`

## What are **`dplyr`** and **`tidyr`**?

The package **`dplyr`** provides easy tools for the most common data
manipulationvtasks. It is built to work directly with data frames, with many
common tasks optimized by being written in a compiled language (C++). An
additional feature is the ability to work directly with data stored in an
external database. The benefits of doing this are that the data can be managed
natively in a relational database, queries can be conducted on that database,
and only the results of the query are returned.

This addresses a common problem with R in that all operations are conducted
in-memory and thus the amount of data you can work with is limited by available
memory. The database connections essentially remove that limitation in that you
can connect to a database of many hundreds of GB, conduct queries on it
directly, and pull back into R only what you need for analysis.

The package **`tidyr`** addresses the common problem of wanting to reshape your
data for plotting and use by different R functions. Sometimes we want data sets
where we have one row per measurement. Sometimes we want a data frame where each
measurement type has its own column, and rows are instead more aggregated groups
(e.g., a time period, an experimental unit like a plot or a batch number).
Moving back and forth between these formats is non-trivial, and **`tidyr`**
gives you tools for this and more sophisticated  data manipulation.

To learn more about **`dplyr`** and **`tidyr`** after the workshop, you may want
to check out this [handy data transformation with **`dplyr`** cheatsheet](https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf)
and this [one about **`tidyr`**](https://github.com/rstudio/cheatsheets/raw/master/data-import.pdf).

In this episode, we will use data on gene expressions stored in the
[TSV (tab-separated values) format](https://en.wikipedia.org/wiki/Tab-separated_values).
We will read in our data using the `read_tsv()` function from the tidyverse
package **`readr`**.

Each line in the file holds the number of reads from are overlapping a
particular gene. Counts are given for 6 individual samples. The dataset consists
of the following columns:

|Â Column   | Description                            |
|----------|----------------------------------------|
| gene     | name of gene                           |
| sample   | 1-letter label for the sample (A to F) |
| genotype | "Wt" (wild type) or "Hom" (homozygous) |
| cnt      | number of reads overlapping the gene   |


```{r, eval=FALSE, purl=TRUE}
download.file(
  url = "https://nbisweden.github.io/module-r-intro-dm-practices/data/gene_counts_r_lesson.csv",
  destfile = "data_raw/gene_counts_r_lesson.csv")
```

```{r, message = FALSE, purl = FALSE}
## load the tidyverse packages, incl. dplyr
library(tidyverse)
```
```{r, eval = FALSE,  purl = FALSE}
gene_counts <- read_tsv("data_raw/gene_counts_r_lesson.tsv")
```
```{r, echo = FALSE, eval = TRUE, purl = FALSE}
# silently read in CSV file with samples
gene_counts <- read_tsv("../data/gene_counts_r_lesson.tsv")
```

```{r, purl = FALSE}
## inspect the data
str(gene_counts)
```

```{r, eval=FALSE, purl=FALSE}
## preview the data
view(gene_counts)
```

Next, we're going to learn some of the most common **`dplyr`** functions:

- `select()`: subset columns
- `filter()`: subset rows on conditions
- `pivot_wider()`: transform rows into columns
- `pivot_longer()`: transform columns into rows
- `mutate()`: create new columns by using information from other columns
- `group_by()` and `summarize()`: create summary statistics on grouped data
- `arrange()`: sort results
- `count()`: count discrete values

## Selecting columns and filtering rows

To select columns of a data frame, use `select()`. The first argument
to this function is the data frame (`gene_counts`), and the subsequent
arguments are the columns to keep.

```{r, results = 'hide', purl = FALSE}
select(gene_counts, gene, sample, cnt)
```

To select all columns *except* certain ones, put a "-" in front of
the variable to exclude it.

```{r, results = 'hide', purl = FALSE}
select(gene_counts, -sample, -genotype)
```

This will select all the variables in `gene_counts` except `sample`
and `genotype`.

To choose rows based on a specific criterion, use `filter()`:

```{r, purl = FALSE}
filter(gene_counts, sample == "A")
```

## Pipes

What if you want to select and filter at the same time? There are three ways to
do this: use intermediate steps, nested functions, or pipes.

With intermediate steps, you create a temporary data frame and use that as input
to the next function, like this:

```{r, purl = FALSE}
gene_counts_A <- filter(gene_counts, sample == "A")
gene_counts_sml <- select(gene_counts_A, gene, genotype, cnt)
```

This is readable, but can clutter up your workspace with lots of objects that
you have to name individually. With multiple steps, that can be hard to keep
track of.

You can also nest functions (i.e. one function inside of another), like this:

```{r, purl = FALSE}
gene_counts_sml <- select(
  filter(gene_counts, sample == "A"), gene, genotype, cnt)
```

This is handy, but can be difficult to read if too many functions are nested, as
R evaluates the expression from the inside out (in this case, filtering, then
selecting).

The last option, *pipes*, are a recent addition to R. Pipes let you take the
output of one function and send it directly to the next, which is useful when
you need to do many things to the same dataset.  Pipes in R look like `%>%` and
are made available via the **`magrittr`** package, installed automatically with
**`dplyr`**. If you use RStudio, you can type the pipe with <kbd>Ctrl</kbd> +
<kbd>Shift</kbd> + <kbd>M</kbd> if you have a PC or <kbd>Cmd</kbd> +
<kbd>Shift</kbd> + <kbd>M</kbd> if you have a Mac.

```{r, purl = FALSE}
gene_counts %>%
  filter(sample == "A") %>%
  select(gene, genotype, cnt)
```

In the above code, we use the pipe to send the `gene_counts` dataset first
through `filter()` to keep rows where `sample` equals `"A"`, then through
`select()` to keep only the `gene`, `genotype`, and `cnt` columns. Since
`%>%` takes the object on its left and passes it as the first argument to the
function on its right, we don't need to explicitly include the data frame as an
argument to the `filter()` and `select()` functions any more.

Some may find it helpful to read the pipe like the word "then". For instance,
in the above example, we took the data frame `gene_counts`, *then* we `filter`ed
for rows with `sample == "A"`, *then* we `select`ed columns `gene`, `genotype`,
and `cnt`. The **`dplyr`** functions by themselves are somewhat simple,
but by combining them into linear workflows with the pipe, we can accomplish
more complex manipulations of data frames.

If we want to create a new object with this smaller version of the data, we
can assign it a new name:

```{r, purl = FALSE}
gene_counts_sml <- gene_counts %>%
  filter(sample == "A") %>%
  select(gene, genotype, cnt)

gene_counts_sml
```

Note that the final data frame is the leftmost part of this expression.

> ## Challenge
>
> Using pipes, subset the `gene_counts` data to include only counts for wild
> type genotypes above 10,000 and retain only the columns `gene` and `cnt`.
>
>> ## Solution
>>
>> ```{r, purl=FALSE}
gene_counts %>%
  filter(cnt > 10000 & genotype == "Wt") %>%
  select(gene, cnt)
```
> {: .solution}
{: .challenge}


### Reshaping with pivot_wider and pivot_longer

You may need to alter the shape of a dataset before you can analyze it.
A general recommendation for tidy data is that you should put each observation 
in its own row, and each variable (i.e. the thing you are measuring) in a
separate column. What you count as an observation depends on the question you
are addressing, and you need to reshape your data accordingly. You can read
more about the principles behind tidy data in Data Carpentry's
[spreadsheet lesson](https://datacarpentry.org/spreadsheet-ecology-lesson/01-format-data/).

There are basically two ways to reshape data frames. One is to turn rows into
columns by transforming values of a variable to column names. The other is to 
turn columns into rows by transforming column names into values of a variable.
We can do these of transformations with the functions `pivot_wider()` and
`pivot_longer()`, respectively.

#### pivot_wider

`pivot_wider()` takes four principal arguments:

1. the data
2. the *id_cols* with the set of variables that uniquely identifies each
   observation.
3. the *names_from* variable to get the output column names from
4. the *values_from* column variable whose values will fill the new column
   variables.


Let's use `pivot_wider()` to transform the gene count data so that we get all
the 6 counts for one gene in the same row.

```{r, purl=FALSE}
gene_counts %>% pivot_wider(
  id_cols = gene,
  names_from = sample,
  values_from = cnt
)
```

This looks pretty good, but we are missing the `genotype` variable. Let's
add this information to the column names as well.

```{r, purl=FALSE}
gene_counts_wide <- gene_counts %>%
  pivot_wider(
    id_cols = gene,
    names_from = c(sample, genotype),
    values_from = cnt
)

str(gene_counts_wide)
```

We store the data frame under the name `gene_counts_wide` so that we can use it
later in this episode.


#### pivot_longer

The opposing situation could occur if we had been provided with data in the
form of `gene_counts_wide`, where the sample labels are column names, but we 
wish to treat them as values of a sample variable instead.

In this situation we are gathering the column names and turning them into a
pair of new variables. One variable represents the column names as values, and
the other variable contains the values previously associated with the column
names.

`pivot_longer()` takes four principal arguments:

1. the data
2. the *cols* with the set of columns we want transform into a variable.
3. the *names_to* column variable we wish to create and fill with values 
associated with the key.
4. the *values_to* column variable we want to fill with the values.

```{r, purl=FALSE}
gene_counts_wide %>% pivot_longer(
    cols = A_Wt:F_Hom,
    names_to = c("sample", "genotype"),
    names_sep = "_",
    values_to = "cnt"
)
```

Since we are transforming column names into two new variables, we must specify a
string separator using the `names_sep` argument.


> ## Challenge
>
> Reshape the `gene_counts` data frame with samples as rows and the three genes
> "Hopx", "Rpl21" and "Calr" as columns.
>
>> ## Solution
>>
>> ```{r, purl = FALSE}
gene_counts %>%
  filter(gene %in% c("Hopx", "Rpl21", "Calr")) %>%
  pivot_wider(
    id_cols = sample,
    names_from = gene,
    values_from = cnt
  )
```
> {: .solution}
{: .challenge}


### Mutate

Frequently you'll want to create new columns based on the values in existing
columns, for example to do unit conversions, or to find the ratio of values in
two columns. For this we'll use `mutate()`.

Let's start by calculating the difference between `D_Hom` and `E_Hom` variables
in the `gene_counts_wide` data frame:

```{r, purl = FALSE}
gene_counts_wide %>%
  mutate(diff_DE = D_Hom - E_Hom)
```

You can also create a second new column based on the first new column within the
same call of `mutate()`:

```{r, purl = FALSE}
gene_counts_wide %>%
  mutate(diff_DE = D_Hom - E_Hom,
         diff_DE_pct = diff_DE / D_Hom * 100)
```

If this runs off your screen and you just want to see the first few rows, you
can use a pipe to view the `head()` of the data. (Pipes work with 
non-**`dplyr`** functions, too, as long as the **`dplyr`** or `magrittr` package
is loaded).

```{r, purl = FALSE}
gene_counts_wide %>%
  mutate(diff_DE = D_Hom - E_Hom,
         diff_DE_pct = diff_DE / D_Hom * 100) %>%
  head()
```

`NaN`s ("Not a Number") will be inserted in the last column whenever we try to
divide a value by zero. If we want to remove those, we can insert a `filter()`
in the chain:

```{r, purl = FALSE}
gene_counts_wide %>%
  mutate(diff_DE = D_Hom - E_Hom,
         diff_DE_pct = diff_DE / D_Hom * 100) %>%
  filter(!is.na(diff_DE_pct)) %>%
  head()
```

`is.na()` is a function that determines whether something is an `NA` (or `NaN`).
The `!` symbol negates the result, so we're asking for every row where weight
*is not* an `NA` (or `NaN`).

> ## Challenge
>
> Create a new data frame from the `gene_counts_wide` data that meets the
> following criteria: contains only the `gene` column and a new column called
> `log_A_Wt` with the common logarithm of the `cnt` values in the `A_Wt` column.
>
> **Hint**: You can use the `log10()` function to calculate the common
> logarithm values.
> 
>> ## Solution
>>
>> ```{r, purl=FALSE}
gene_counts_wide %>%
  mutate(log_A_Wt = log10(A_Wt)) %>%
  select(gene, log_A_Wt)
```
> {: .solution}
{: .challenge}


### Split-apply-combine data analysis and the `summarize()` function

Many data analysis tasks can be approached using the *split-apply-combine*
paradigm: split the data into groups, apply some analysis to each group, and
then combine the results. **`dplyr`** makes this very easy through the use of
the `group_by()` function.


#### The `summarize()` function

`group_by()` is often used together with `summarize()`, which collapses each
group into a single-row summary of that group. `group_by()` takes as arguments
the column names that contain the **categorical** variables for which you want
to calculate the summary statistics. So to compute the mean `weight` by sex:

```{r, purl = FALSE}
gene_counts %>%
  group_by(genotype) %>%
  summarize(mean = mean(cnt))
```

You can also group by multiple columns:

```{r, purl = FALSE}
gene_counts %>%
  group_by(gene, genotype) %>%
  summarize(mean = mean(cnt)) %>% 
  head()
```

Once the data are grouped, you can also summarize multiple variables at the same
time (and not necessarily on the same variable). For instance, we could add a
column indicating the minimum count for each genotype for each gene:

```{r, purl = FALSE}
gene_counts %>%
  group_by(gene, genotype) %>%
  summarize(mean = mean(cnt),
            min = min(cnt)) %>%
  head()
```

It is sometimes useful to rearrange the result of a query to inspect the values.
For instance, we can sort on `min_cnt` to put the lowest numbers first:


```{r, purl = FALSE}
gene_counts %>%
  group_by(gene, genotype) %>%
  summarize(mean = mean(cnt),
            min = min(cnt)) %>%
  arrange(min)
```

To sort in descending order, we need to add the `desc()` function. If we want
to sort the results by decreasing order of mean count:

```{r, purl = FALSE}
gene_counts %>%
  group_by(gene, genotype) %>%
  summarize(mean = mean(cnt),
            min = min(cnt)) %>%
  arrange(desc(min))
```

#### Counting

When working with data, we often want to know the number of observations found
for each factor or combination of factors. For this task, **`dplyr`** provides
`count()`. For example, if we wanted to count the number of rows of data for
each sample, we would do:

```{r, purl = FALSE}
gene_counts %>%
    count(sample) 
```

The `count()` function is shorthand for something we've already seen: grouping
by a variable, and summarizing it by counting the number of observations in that
group. In other words, `gene_counts %>% count(sample)` is equivalent to:  

```{r, purl = FALSE}
gene_counts %>%
    group_by(sample) %>%
    summarize(n = n())
```

We can also combine `count()` with other functions such as `filter()`. Here
we will count for each sample the number of genes with more than 10,000 reads.

```{r, purl = FALSE}
gene_counts %>%
    filter(cnt > 10000) %>%
    count(gene) 
```

The example above shows the use of `count()` to count the number of
rows/observations for *one* factor (i.e., `gene`). If we wanted to count
*combination of factors*, such as `gene` and `genotype`, we would specify the
first and the second factor as the arguments of `count()`:

```{r purl = FALSE}
gene_counts %>%
  filter(cnt > 10000) %>%
  count(gene, genotype) 
```

With the above code, we can proceed with `arrange()` to sort the table 
according to a number of criteria so that we have a better comparison. 
For instance, we might want to arrange the table above in (i) an alphabetical
order of the levels of the genotypes and (ii) in descending order of the count:

```{r purl = FALSE}
gene_counts %>%
  filter(cnt > 10000) %>%
  count(gene, genotype) %>%
  arrange(genotype, desc(n))
```

From the table above, we may learn that, for instance, there are 75 observations of 
the *albigula* species that are not specified for its sex (i.e. `NA`).


> ## Challenge
>
> * For each sample in the `gene_counts` data frame, how many of the genes have
>   zero reads?
>
>> ## Solution
>>
>> ```{r, purl=FALSE}
gene_counts %>%
  filter(cnt == 0) %>%
  count(gene) 
```
> {: .solution}
>
> * Use `group_by()` and `summarize()` to find the mean and standard deviation
>   of the number of reads for each genotype and gene. Then reshape the data
>   so that the mean and standard deviation for for the two genotypes are
>   displayed in separate columns.
>
>   **Hint:** calculate the standard deviation with the `sd()` function.
>
>> ## Solution
>>
>> ```{r, purl=FALSE}
gene_counts %>%
    group_by(gene, genotype) %>%
    summarize(
        mean = mean(cnt),
        stdev = sd(cnt)
    ) %>% 
    pivot_wider(
        id_cols = gene,
        names_from = genotype,
        values_from = (c(mean, stdev))
    )
```
> {: .solution}
{: .challenge}


# Exporting data

Now that you have learned how to use **`dplyr`** to extract information from
or summarize your raw data, you may want to export these new data sets to share
them with your collaborators or for archival.

Similar to the `read_csv()` function used for reading CSV files into R, there is
a `write_csv()` function that generates CSV files from data frames.

Before using `write_csv()`, we are going to create a new folder, `data`,
in our working directory that will store this generated dataset. We don't want
to write generated datasets in the same directory as our raw data. It's good
practice to keep them separate. The `data_raw` folder should only contain the
raw, unaltered data, and should be left alone to make sure we don't delete or
modify it. In contrast, our script will generate the contents of the `data`
directory, so even if the files it contains are deleted, we can always
re-generate them.

To write the `gene_counts_wide` data frame that we prepared earlier to a file,
we can do the following:

```{r, purl=FALSE, eval=FALSE}
write_csv(gene_counts_wide, file = "data/gene_counts_wide.csv")
```
