---
title: Exercises
teaching: 0
exercises: 0
questions:
- ""
objectives:
- ""
keypoints:
- ""
source: Rmd
---

```{r chunk_options, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
```

## 1. Introducing R and RStudio IDE

> ## Discussion: Your experience and expectations
>
> * What has motivated you to learn R? Have you had a research question for which
>   spreadsheet programs such as Excel have proven difficult to use, or where the
>   size of the data set created issues?
> * In what situations do you imagine that using R will be most valuable to you? 
{: .discussion}


## 2. R Basics

> ## Challenge 2.1
>
> What are the values after each statement in the following?
>
> ```{r, purl = FALSE}
mass <- 47.5            # mass?
age  <- 122             # age?
mass <- mass * 2.0      # mass?
age  <- age - 20        # age?
mass_index <- mass/age  # mass_index?
```
{: .challenge}


> ## Challenge 2.2
>
> Write a function that returns the sum of two arguments given to it.
>
>> ## Solution
>>
>> ```{r, purl = FALSE}
# Calculate the sum of x and y
add_two_args <- function(x, y) { x + y }

# Call the function
add_two_args(9, 7)
```
>>
> {: .solution}
{: .challenge}


> ## Challenge 2.3
>
> * We’ve seen that atomic vectors can be of type character,
>   numeric (or double), integer, and logical. But what happens if we try to mix
>   these types in a single vector?
> 
>> ## Answer
>>
>> R implicitly converts them to all be the same type
> {: .solution}
>
> * What will happen in each of these examples? (hint: use `class()`
>   to check the data type of your objects):
>
>     ```r
>     num_char <- c(1, 2, 3, "a")
>     num_logical <- c(1, 2, 3, TRUE)
>     char_logical <- c("a", "b", "c", TRUE)
>     tricky <- c(1, 2, 3, "4")
>     ```
>
> * Why do you think it happens?
>
>> ## Answer
>>
>> Vectors can be of only one data type. R tries to convert (coerce) the content
>> of this vector to find a "common denominator" that doesn't lose any
>> information.
> {: .solution}
>
> * How many values in `combined_logical` are `"TRUE"` (as a character) in the
>   following example (reusing the 2 `..._logical`s from above):
>
>     ```r
>     combined_logical <- c(num_logical, char_logical)
>     ```
>
>> ## Answer
>>
>> Only one. There is no memory of past data types, and the coercion happens the
>> first time the vector is evaluated. Therefore, the `TRUE` in `num_logical`
>> gets converted into a `1` before it gets converted into `"1"` in
>> `combined_logical`.
> {: .solution}
>
> * You've probably noticed that objects of different types get
>   converted into a single, shared type within a vector. In R, we
>   call converting objects from one class into another class
>   _coercion_. These conversions happen according to a hierarchy,
>   whereby some types get preferentially coerced into other
>   types. Can you draw a diagram that represents the hierarchy of how
>   these data types are coerced?
>
>> ## Answer
>>
>> logical &#8594; numeric &#8594; character &#8592; logical
> {: .solution}
{: .challenge}


> ## Challenge 2.4 (optional)
>
> * Can you figure out why `"four" > "five"` returns `TRUE`?
> 
>> ## Answer
>>
>> When using ">" or "<" on strings, R compares their alphabetical order. 
>> Here "four" comes after "five", and therefore is "greater than" it.
> {: .solution}
{: .challenge}


> ## Challenge 2.5
>
> 1. Using this vector of heights in inches, create a new vector,
>    `heights_no_na`, with the NAs removed.
>
>     ```r
>     heights <- c(63, 69, 60, 65, NA, 68, 61, 70, 61, 59, 64, 69, 63, 63, NA, 72, 65, 64, 70, 63, 65)
>     ```
> 2. Use the function `median()` to calculate the median of the `heights` vector.
>
> 3. Use R to figure out how many people in the set are taller than 67 inches.
>
>> ## solution
>>
>> ```{r}
heights <- c(63, 69, 60, 65, NA, 68, 61, 70, 61, 59, 64, 69, 63, 63, NA, 72, 65, 64, 70, 63, 65)

# 1.
heights_no_na <- heights[!is.na(heights)] 
# or
heights_no_na <- na.omit(heights)
# or
heights_no_na <- heights[complete.cases(heights)]

# 2.
median(heights, na.rm = TRUE)

# 3.
heights_above_67 <- heights_no_na[heights_no_na > 67]
length(heights_above_67)
```
> {: .solution}
{: .challenge}


## 3. R Basics continued - factors and data frames

### Loading the samples data

```{r, eval=FALSE, purl=TRUE}
download.file(
  url = "https://nbisweden.github.io/module-r-intro-dm-practices/data/samples_r_lesson.csv",
  destfile = "data_raw/samples_r_lesson.csv")
```

```{r, message = FALSE, purl = FALSE}
## load the tidyverse packages, incl. dplyr
library(tidyverse)
```

```{r, eval = FALSE,  purl = FALSE}
samples <- read_csv("data_raw/samples_r_lesson.csv")
```
```{r, echo = FALSE, eval = TRUE, purl = FALSE}
# silently read in CSV file with samples
samples <- read_csv("../data/samples_r_lesson.csv")
```

### Inspecting data frames

* Size:
    * `dim(samples)` - returns a vector with the number of rows in the first
      element, and the number of columns as the second element
      (the **dim**ensions of the object)
    * `nrow(samples)` - returns the number of rows
    * `ncol(samples)` - returns the number of columns

* Content:
    * `head(samples)` - shows the first 6 rows
    * `tail(samples)` - shows the last 6 rows

* Names:
    * `names(samples)` - returns the column names (synonym of `colnames()` for
      `data.frame` objects)
    * `rownames(samples)` - returns the row names

* Summary:
    * `str(samples)` - structure of the object and information about the class,
      length and content of each column
    * `summary(samples)` - summary statistics for each column


> ## Challenge 3.1
>
> Based on the output of `str(samples)`, can you answer the following questions?
>
> * What is the class of the object `samples`?
> * How many rows and how many columns are in this object?
>
>> ## Solution
>>
>> ```{r, purl=FALSE}
str(samples)
```
>> * The object `samples` is of class `data.frame`, or more specifically a
>>   `tibble` (`spec_tbl_df/tbl_df/tbl/data.frame`)
>> * Rows and columns: 100 rows and 12 columns
> {: .solution}
{: .challenge}


> ## Challenge 3.2
>
> 1. Create a `data.frame` (`samples_20`) containing only the data in
>    row 20 of the `samples` dataset.
>
> 2. Notice how `nrow()` gave you the number of rows in a `data.frame`?
>
>      * Use that number to pull out just that last row in the data frame.
>      * Compare that with what you see as the last row using `tail()` to make
>        sure it's meeting expectations.
>      * Pull out that last row using `nrow()` instead of the row number.
>      * Create a new data frame (`samples_last`) from that last row.
>
> 3. Use `nrow()` to extract the row that is in the middle of the data
>    frame. Store the content of this row in an object named `samples_middle`.
>
> 4. Combine `nrow()` with the `-` notation above to reproduce the behavior of
>    `head(samples)`, keeping just the first through 6th rows of the samples
>    dataset.
>
>> ## Solution
>>
>>```{r, purl=FALSE}
## 1.
samples_20 <- samples[20, ]
## 2.
# Saving `n_rows` to improve readability and reduce duplication
n_rows <- nrow(samples)
samples_last <- samples[n_rows, ]
## 3.
samples_middle <- samples[n_rows / 2, ]
## 4.
samples_head<- samples[-(7:n_rows), ]
```
> {: .solution}
{: .challenge}


> ## Challenge 3.3
>
> 1. Change the columns `strain` and `organism_part` in the `samples`
>    data frame into a factor.
>
> 2. Using the functions you have learnt so far, can you find out...
>
>      * How many samples were taken from brain tissue?
>      * How many different mouse strains are in the `strain` column?
>
>> ## Solution
>>
>> ```{r, purl=FALSE}
samples$strain <- factor(samples$strain)
samples$organism_part <- factor(samples$organism_part)
summary(samples$organism_part)
nlevels(samples$strain)
```
>>
>> * How many brain samples? There are 16 "brain" values in the `organism_part`
     column.
>> * how many mouse strains? There are 2 strains in the `strain` column.
> {: .solution}
{: .challenge}

> ## Challenge 3.4
>
> 1. We have seen how data frames are created when using `read_csv()`, but they
>    can also be created by hand with the `data.frame()` function.  There are a
>    few mistakes in this hand-crafted `data.frame`. Can you spot and fix them?
>    Don't hesitate to experiment!
>
>    ```{r, eval=FALSE, purl=FALSE}
>    animal_data <- data.frame(
>              animal = c(dog, cat, sea cucumber, sea urchin),
>              feel = c("furry", "squishy", "spiny"),
>              weight = c(45, 8 1.1, 0.8)
>              )
>    ```
>
> 2. Can you predict the class for each of the columns in the following example?
>    Check your guesses using `str(country_climate)`:
>      * Are they what you expected? Why? Why not?
>      * What would you need to change to ensure that each column had the
>        accurate data type?
>
>     ```{r, eval=FALSE, purl=FALSE}
>     country_climate <- data.frame(
>            country = c("Canada", "Panama", "South Africa", "Australia"),
>            climate = c("cold", "hot", "temperate", "hot/temperate"),
>            temperature = c(10, 30, 18, "15"),
>            northern_hemisphere = c(TRUE, TRUE, FALSE, "FALSE"),
>            has_kangaroo = c(FALSE, FALSE, FALSE, 1)
>            )
>     ```
>
>> ## Solution
>>
>> * missing quotations around the names of the animals
>> * missing one entry in the `feel` column (probably for one of the furry animals)
>> * missing one comma in the `weight` column
>> * `country`, `climate`, `temperature`, and `northern_hemisphere` are
>>    characters; `has_kangaroo` is numeric
>> * using `factor()` one could replace character columns with factors columns
>> * removing the quotes in `temperature` and `northern_hemisphere` and replacing 1
>>   by TRUE in the `has_kangaroo` column would give what was probably 
>>   intended
> {: .solution}
{: .challenge}


## 4. Aggregating and Analyzing Data with dplyr

### Load the gene counts dataset


| Column   | Description                            |
|----------|----------------------------------------|
| gene     | name of gene                           |
| sample   | 1-letter label for the sample (A to F) |
| genotype | "Wt" (wild type) or "Hom" (homozygous) |
| cnt      | number of reads overlapping the gene   |


```{r, eval=FALSE, purl=TRUE}
download.file(
  url = "https://nbisweden.github.io/module-r-intro-dm-practices/data/gene_counts_r_lesson.csv",
  destfile = "data_raw/gene_counts_r_lesson.csv")
```

```{r, message = FALSE, purl = FALSE}
## load the tidyverse packages, incl. dplyr
library(tidyverse)
```
```{r, eval = FALSE,  purl = FALSE}
cnts <- read_tsv("data_raw/gene_counts_r_lesson.tsv")
```
```{r, echo = FALSE, eval = TRUE, purl = FALSE}
# silently read in TSV file with gene counts
cnts <- read_tsv("../data/gene_counts_r_lesson.tsv")
```

- `select()`: subset columns
- `filter()`: subset rows on conditions
- `pivot_wider()`: transform rows into columns
- `pivot_longer()`: transform columns into rows
- `mutate()`: create new columns by using information from other columns
- `group_by()` and `summarize()`: create summary statistics on grouped data
- `arrange()`: sort results
- `count()`: count discrete values

> ## Challenge 4.1
>
> Using pipes, subset the `cnts` data to include only counts for wild type
> genotypes above 10,000 and retain only the columns `gene` and `cnt`.
>
>> ## Solution
>>
>> ```{r, purl=FALSE}
cnts %>%
  filter(cnt > 10000 & genotype == "Wt") %>%
  select(gene, cnt)
```
> {: .solution}
{: .challenge}


> ## Challenge 4.2
>
> Reshape the `cnts` data frame with samples as rows and the three genes
> "Hopx", "Rpl21" and "Calr" as columns.
>
>> ## Solution
>>
>> ```{r, purl = FALSE}
cnts %>%
  filter(gene %in% c("Hopx", "Rpl21", "Calr")) %>%
  pivot_wider(
    id_cols = sample,
    names_from = gene,
    values_from = cnt
  )
```
> {: .solution}
{: .challenge}

> ## Challenge 4.3
>
> Create a new data frame from the `cnts_wide` data that meets the
> following criteria: contains only the `gene` column and a new column called
> `log_A_Wt` with the common logarithm of the `cnt` values in the `A_Wt` column.
>
> **Hint**: You can use the `log10()` function to calculate the common
> logarithm values.
> 
>> ## Solution
>>
>> ```{r, purl=FALSE}
cnts_wide %>%
  mutate(log_A_Wt = log10(A_Wt)) %>%
  select(gene, log_A_Wt)
```
> {: .solution}
{: .challenge}


> ## Challenge 4.4
>
> * For each sample in the `cnts` data frame, how many of the genes have
>   zero reads?
>
>> ## Solution
>>
>> ```{r, purl=FALSE}
cnts %>%
  filter(cnt == 0) %>%
  count(gene) 
```
> {: .solution}
>
> * Use `group_by()` and `summarize()` to find the mean and standard deviation
>   of the number of reads for each genotype and gene. Then reshape the data
>   so that the mean and standard deviation for for the two genotypes are
>   displayed in separate columns.
>
>   **Hint:** calculate the standard deviation with the `sd()` function.
>
>> ## Solution
>>
>> ```{r, message = FALSE, purl=FALSE}
cnts %>%
    group_by(gene, genotype) %>%
    summarize(
        mean = mean(cnt),
        stdev = sd(cnt)
    ) %>% 
    pivot_wider(
        id_cols = gene,
        names_from = genotype,
        values_from = (c(mean, stdev))
    )
```
> {: .solution}
{: .challenge}


### Exporting data to be used in later exercises

```{r, purl=FALSE, eval=FALSE}
# Create a vector with a subset of genes
genes_subset <- c(
  "Rbpj", "Nop10", "Nsun2", "Cxcl5", "Cfdp1", "Furin", "Adcy8", "Grin1",
  "Itpka", "Ippk", "Aff2", "Rab43", "Ofd1", "Edem2", "Xlr4c", "Mpc2",
  "Ect2l", "Pcgf1", "Nov", "Sys1", "Ackr2", "Acat1", "Phkg2", "Hecw2",
  "Ctps", "Arl5b", "Wnt3a", "Evc", "Wibg", "Axl", "Usp19", "Agbl2",
  "Eif3f", "Mark4", "Sbk2", "Eri3", "Tab1", "Cd276", "Sox9", "Ces2e",
  "Htra3", "Foxj2", "Arl4d", "Wdr1", "Taco1", "Aebp2", "Nemf", "Fpgt",
  "Helq", "Nudt6")

# Filter out subset of genes
cnts_sml_wide <- cnts_wide %>%
  filter(gene %in% genes_subset)

cnts_sml_long <- cnts %>%
  filter(gene %in% genes_subset)

# Write data frame to CSV
write_csv(cnts_sml_wide, file = "gene_counts_sml_wide.csv")

# Write data frame to CSV
write_csv(cnts_sml_long, file = "gene_counts_sml_long.csv")
```

## 5. Data Visualization with ggplot2

```{r load-package, message=FALSE, purl=FALSE}
library(tidyverse)
```

### Load the small gene counts datasets

If not still in the workspace, load the two small datasets that we created in
the previous episode.

```{r load-data, eval = FALSE, purl = FALSE}
cnts_sml_long <- read_csv("gene_counts_sml_long.csv")
cnts_sml_wide <- read_csv("gene_counts_sml_wide.csv")
```

```{r, echo = FALSE, eval = TRUE, include = FALSE, purl = FALSE}
## silently read in CSV files
cnts_sml_long <- read_csv("../data/gene_counts_sml_long.csv")
cnts_sml_wide <- read_csv("../data/gene_counts_sml_wide.csv")
```

To make the `cnts_sml_wide` dataset a bit more interesting for the plotting in
this episode, we will add a fictional column as a factor that we call
`gene_category`:

```{r purl = FALSE}
cnts_sml_wide$gene_category = factor(c(
    2, 2, 1, 1, 1, 3, 3, 1, 1, 1, 3, 3, 1, 3, 3, 2, 2, 3, 3, 2, 2, 2, 2, 3, 2,
    1, 2, 3, 2, 3, 1, 3, 1, 2, 3, 2, 1, 2, 3, 1, 1, 1, 3, 2, 1, 2, 2, 3, 3, 1))

head(cnts_sml_wide)
```


> ## Challenge 5.1 (optional)
>
> Scatter plots can be useful exploratory tools for small datasets. For data
> sets with large numbers of observations, overplotting of points can be a
> limitation of scatter plots. One strategy for handling such settings is to use
> hexagonal binning of observations. The plot space is tessellated into
> hexagons. Each hexagon is assigned a color based on the number of observations
> that fall within its boundaries. To use hexagonal binning with **`ggplot2`**, 
> irst install the R package `hexbin` from CRAN:
>
>
> ```{r, eval = FALSE}
  install.packages("hexbin")
  library(hexbin)
```
>
> Then use the `geom_hex()` function:
>
> ```{r, eval = FALSE}
  cnts_plot +
  geom_hex()
```
>
> - What are the relative strengths and weaknesses of a hexagonal bin plot
>   compared to a scatter plot? Examine the above scatter plot and compare it
>   with the hexagonal bin plot that you created.
```
{: .challenge}

> ## Challenge 5.2
>
> Use what you just learned to create a scatter plot the gene category against
> the counts for sample `A_Wt`. Is this a good way to show this type of data?
>
>> ## Solution
>>
>> ```{r scatter-challenge-answer, answer=TRUE, purl=FALSE}
ggplot(data = cnts_sml_wide, 
       mapping = aes(x = gene_category, y = A_Wt)) +
  geom_point(aes(color = gene_category))
```
>> Since we have counts that varies over a wide range, one could for example
>> use a logarithmic y-scale instead:
>>
>> ```{r scatter-challenge-answer2, answer=TRUE, purl=FALSE}
ggplot(data = cnts_sml_wide, 
       mapping = aes(x = gene_category, y = A_Wt)) +
  geom_point(aes(color = gene_category)) +
  scale_y_log10()
```
> {: .solution}
{: .challenge}

> ## Challenges 5.3
>
> Use the code below, and try to incrementally make changes to the plot.
>
>```{r box-with-points-challenge, eval=FALSE, echo=TRUE, purl=FALSE}
ggplot(data = cnts_sml_long, mapping = aes(x = sample, y = cnt)) +
    geom_boxplot(alpha = 0) +
    geom_jitter(alpha = 0.5, color = "tomato")
```
>
> Boxplots are useful summaries, but hide the *shape* of the distribution. For
> example, if there is a bimodal distribution, it would not be observed with a
> boxplot. An alternative to the boxplot is the violin plot (sometimes known as 
> a beanplot), where the shape (of the density of points) is drawn.
>
> * Replace the box plot with a violin plot; see `geom_violin()`.
>
>> ## Solution
>>
>> ```{r violinplot-with-points, eval=FALSE, echo=TRUE, purl=FALSE}
# Violin plot instead of boxplot
ggplot(data = cnts_sml_long, mapping = aes(x = sample, y = cnt)) +
    geom_boxplot(alpha = 0) +
    geom_jitter(alpha = 0.5, color = "tomato")
```
> {: .solution}
>
> In many types of data, it is important to consider the *scale* of the
> observations. For example, it may be worth changing the scale of the axis to
> better distribute the observations in the space of the plot. Changing the
> scale of the axes is done similarly to adding/modifying other components:
>
> * Represent the read counts on the log~10~ scale; see `scale_y_log10()`.
>
>> ## Solution
>>
>>```{r violinplot-logscale-answer, purl=FALSE}
# Violin plot with log-scale
ggplot(data = cnts_sml_long, mapping = aes(x = sample, y = cnt)) +
  geom_violin(alpha = 0) +
  geom_jitter(alpha = 0.5, color = "tomato") +
  scale_y_log10()
```
> {: .solution}
>
> * We've now looked at the distribution of the counts within samples. Try
>   making a new plot to explore the distribution for each `genotype` instead.
>   Overlay the boxplot layer on a jitter layer to show actual counts.
>
>> ## Solution
>>
>> ```{r violinplot-by-genotype-answer, purl=FALSE}
# Violin plot by genotype
ggplot(data = cnts_sml_long, mapping = aes(x = genotype, y = cnt)) +
  geom_violin(alpha = 0) +
  geom_jitter(alpha = 0.5, color = "tomato") +
  scale_y_log10()
```
> {: .solution}
>
> * Add color to the data points on your boxplot according to the sample's
>   genotype of the actual sample.
>
>> ## Solution
>>
>> ```{r violinplot-by-genotype-color-answer, purl=FALSE}
# Violin plot by genotype in color
ggplot(data = cnts_sml_long, mapping = aes(x = genotype, y = cnt)) +
  geom_violin(alpha = 0) +
  geom_jitter(alpha = 0.5, aes(color = genotype)) +
  scale_y_log10()
```
> {: .solution}
>
> * Reorder the genotype categories so that "Wt" is shown to the left and "Hom"
>   to the right (**hint:** use the `factor()` function).
>
>> ## Solution
>>
>> ```{r violinplot-by-genotype-ordering-answer, purl=FALSE}
# Violin plot by reordered genotypes

# Create a factor with reordered levels
cnts_sml_long$genotype <- factor(cnts_sml_long$genotype, levels = c("Wt", "Hom"))

# Now create the plot in the same way as before
ggplot(data = cnts_sml_long, mapping = aes(x = genotype, y = cnt)) +
  geom_violin(alpha = 0) +
  geom_jitter(alpha = 0.5, aes(color = genotype)) +
  scale_y_log10()
```
> {: .solution}
{: .challenge}

> ## Challenge 5.4
>
> Use what you just learned to create a plot that show how the mean count for
> each genotype differ between the three genes "Evc", "Mpc" and "Nemf".
>
>> ## Solution
>> 
>> ```{r average-gene-counts, message=FALSE, purl=FALSE}
cnts_sml_long %>%
  filter(gene %in% c("Evc", "Mpc2", "Nemf")) %>%
  group_by(gene, genotype) %>%
  summarize(mean=mean(cnt)) %>%
  ggplot(aes(x = genotype, y = mean)) +
    geom_bar(stat = "identity") +
    facet_wrap(vars(gene))
```
> {: .solution}
{: .challenge}

> ## Challenge 5.5
>
> With all of this information in hand, please take another five minutes to either
> improve one of the plots generated in this exercise or create a beautiful graph
> of your own. Use the RStudio [**`ggplot2`** cheat sheet](https://www.rstudio.com/wp-content/uploads/2016/11/ggplot2-cheatsheet-2.1.pdf) 
> for inspiration.
>
> Here are some ideas:
>
> * See if you can change the thickness of the lines.
> * Can you find a way to change the name of the legend? What about its labels?
> * Try using a different color palette (see
>   [http://www.cookbook-r.com/Graphs/Colors_(ggplot2)/]).
{: .challenge


### Exporting plots

```{r ggsave-example, eval = FALSE, purl = FALSE}
cnts_plot <- cnts_sml_long %>%
  # Filter out three genes
  filter(gene %in% c("Evc", "Mpc2", "Nemf")) %>%
  # Create a separate barplot for each gene
  ggplot(aes(x = sample, y = cnt, fill = genotype)) +
    geom_bar(stat = "identity") +
    facet_wrap(facets = vars(gene)) +
    labs(title = "Number of reads for selected genes",
         x = "Sample",
         y = "Number of reads") +
    scale_fill_discrete(
      name = "Genotype",
      labels = c("Wild type", "Homozygous")) + 
    theme_bw() +
    theme(text = element_text(size = 16))

# Save the file
ggsave("gene_counts_plot.png", cnts_plot, width = 15, height = 10)
```


## 6. Producing Reports With knitr

> ## Challenge 6.1
>
> Create a new R Markdown document. Delete all of the R code chunks
> and write a bit of Markdown (some sections, some italicized
> text, and an itemized list).
>
> Convert the document to a webpage.
{: .challenge}

> ## Challenge 6.2
>
> Add code chunks to:
>
> - Load the `tidyverse` and `ggplot2` packages
> - Read the data from the file `gene_counts_sml_long.csv` (you should have
>   the file in your working directory from earlier episodes).
> - Create a plot based on the loaded data.
{: .challenge}

> ## Challenge 6.3
>
> Use chunk options to control the size of a figure and to hide the
> code.
{: .challenge}

> ## Challenge 6.4
>
> Try out a bit of in-line R code.
{: .challenge}
